# Sex-biased and tissue-biased gene expression evolution in _C. remanei_ and _C. latens_

## Author
Athmaja Viswanath <sup>1</sup>1, Daniel D. Fusca <sup>1</sup>, John A. Calarco <sup>2</sup>, and Asher D. Cutter <sup>1</sup><sup>*</sup>

<sup>1</sup>Department of Ecology & Evolutionary Biology
<sup>2</sup>Department of Cell & Systems Biology
University of Toronto, Toronto, Ontario, M5S3B2 Canada

<sup>*</sup>Author for correspondence: asher.cutter@utoronto.ca

**Athmaja Viswanath**  
Email: [athmaja96@gmail.com](mailto:athmaja96@gmail.com)  

## Abstract
The regulation of gene expression is crucial for an organismâ€™s development. Differential gene regulation enables precise spatio-temporal control of gene expression across developmental stages, tissues, and sexes, contributing to the proper development of the organism. As species evolve, mutational changes can lead to divergence in gene regulation, leading to species-specific gene expression patterns. Divergence in gene regulation can specifically drive evolutionary changes to sex-biased and tissue-biased genetic networks between species, leading to species-specific expression. Despite extensive research, the most prevalent mechanisms that underpin divergence in tissue-dependent gene expression between species remain unresolved. Here I study the role of sex-biased and tissue-biased gene regulatory changes in producing species-specific gene expression patterns. I analyzed transcriptomes from two tissues (gonad and soma) for each of two sexes (male and female) from the two closest known sister species of _Caenorhabditis_ nematodes (_C. remanei_ and _C. latens_). I used differential gene expression analysis and gene co-expression network analysis to identify collections of genes with tissue-biased and sex-biased gene expression differences within and between species. My analysis provides evidence for widespread gene expression divergence between these morphologically and developmentally cryptic sister species. Genes showing conserved expression between tissues and sexes outnumber genes with species differences, implicating shared selection pressures of different tissues and sexes as outweighing species-specific selection and developmental system drift in shaping overall transcriptome profiles. Although tissue-biased expression profiles are mostly conserved between species, I find that sex-biased genes, particularly male-biased genes, contribute disproportionately to species-differences in gene expression, consistent with a disproportionate role for male-biased selection driving gene expression divergence.

## Repository Overview
On-going analysis for _C. remanei_ and _C. latens_ datasets. This dataset contains data for two species (_C. remanei_ & _C. latens_), sex (males & females) and tissues (gonad & soma)

## Repository Structure

### 1-Input
- Input data files currently not made available

### 2-Processing
- Contains scripts for Differential Gene Expression Analysis and Weighted gene co-expression netwrok analysis (WGCNA)
- Scripts are part of on-going analyses and are subject to change

### 3-Output
- Contains all the generated figures and additional files resulting from the analyses
