---
title: "1.DGEAnalysis"
author: "Athmaja Viswanath"
date: "2024-09-19"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
```

#0. LOADING REQUIRED PACKAGES
```{r loading packages}
library(DESeq2)
library(ggplot2)
library(cowplot) #add on to ggplot for better themes and figure customization
library(lemon) #to work with legends and axes in ggplot2
library(dplyr)
library(gdata)
library(RColorBrewer)
library(colorBlindness)
library(colorspace)
library(tidyverse)

theme_set(theme_classic())
```

#1.Data preparation 
##A. Separating orthologous gene names so that they can be used to filter readcounts for each species.
Total number of orthologous genes detected is 13785 genes across C. remanei and C. latens (gene list from Daniel)
```{r}
# Load ortholog data from a tab-separated file, ignoring comments
orthologs = read.table("../1-Input/new_1to1_orthologgenelist.txt", sep="\t", header=TRUE, comment.char="#")

# Extract the relevant columns into separate variables
cre_ortho = orthologs[, 1]            # First column: cre_ortho, Cre names
clat_ortho = orthologs[, 2]           # Second column: clat_ortho, Clat names
cumulative_ortho = orthologs[, 3]     # Third column: cumulative_ortho, joint names

# Display the first few rows of the data frames
head(orthologs)
head(cre_ortho)
head(clat_ortho)

##Number of genes 
nrow(orthologs)
length(cre_ortho)
length(clat_ortho)
```

##B Filtering readcoutns for all samples to retain orthologous genes and combining to make one readcount file
Total number of genes in C. remanei = 20091
Total number of gene sin C. latens = 23073
Total number of 1:! orthologous genes = 13785

```{r}
#########
#C.remanei
#########
# Load counts for C. remanei
cre_counts = read.table("../1-Input/C.remanei_cumulative.txt", sep="\t", header=TRUE, row.names=1, comment.char="#")
# Look at the structure of data
head(cre_counts)
colnames(cre_counts)
nrow(cre_counts) #20091
# Filter to include only orthologs and exclude Whole animal samples
cre_counts_ortho = cre_counts[cre_ortho, -c(13:15)] 
# Update row names to match cumulative orthologs
rownames(cre_counts_ortho) = cumulative_ortho 
# Check the number of rows and column names
nrow(cre_counts_ortho)
colnames(cre_counts_ortho)

#########
#C.latens
#########
# Load counts for C. latens
clat_counts = read.table("../1-Input/C.latens_cumulative.txt", sep="\t", header=TRUE, row.names=1, comment.char="#")
# Look at the structure of data
head(clat_counts)
colnames(clat_counts)
nrow(clat_counts) #23073
# Filter to include only orthologs and exclude Whole animal samples
clat_counts_ortho = clat_counts[clat_ortho, -c(13:15)]
# Update row names to match cumulative orthologs
rownames(clat_counts_ortho) = cumulative_ortho
# Check the number of rows and column names for C. latens
nrow(clat_counts_ortho)
names(clat_counts_ortho)


# Combine ortholog counts from C. remanei and C. latens
wt_counts_orthologs = cbind(cre_counts_ortho, clat_counts_ortho)

# Display current column names
colnames(wt_counts_orthologs)

# Clean up column names by removing the "X" prefix
colnames(wt_counts_orthologs) = gsub("X", "", colnames(wt_counts_orthologs))
colnames(wt_counts_orthologs)
```

#2 PERFORMING DIFFERNTIAL GENE EXPRESSION ANALYSIS
##A Preparing coldata
```{r}
# Prepare the coldata data frame
sample_name = colnames(wt_counts_orthologs)  # Get sample names
tissue = substr(sample_name, 3, 3)            # Extract tissue type from sample names
sex = substr(sample_name, 2, 2)               # Extract sex from sample names
species = c(rep("Cre", 12), rep("Clat", 12))  # Assign species labels
batch = rep(c(1, 2, 3),8)                # Assign batch numbers

# Create the coldata data frame
coldata = data.frame(sample_name, species, sex, tissue, batch)

# Convert columns to factors for analysis
coldata$species = factor(coldata$species)
coldata$tissue = factor(coldata$tissue)
coldata$sex = factor(coldata$sex)
coldata$batch = factor(coldata$batch)

# Display the coldata data frame
coldata  # Note: WM samples are not included
```

##B Performing differential gene expression using DESeq2 and extracting model matrix
```{r}
#Run deseq2 analysis 
dds_wt = DESeqDataSetFromMatrix(
  countData = wt_counts_orthologs, 
  colData = coldata, 
  design = ~ species + tissue + sex + species:tissue + sex:tissue + species:sex)

# Perform the DESeq analysis
dds_wt = DESeq(dds_wt)

# Display the names of the results generated by the model
resultsNames(dds_wt)

# Create the model matrix based on the design
mod_mat_wt <- model.matrix(design(dds_wt), colData(dds_wt))
```

##C Extracting constrasts/comparisons from the main model

###I BETWEEN SPECIES
Identifying species-biased genes (INCLUDING THOSE THAT ARE DUE TO INTERATION TERM)

Results: From 13,503 gene orthologs with non-zero read counts, we found ~49% of genes (nDEG = 6563) to show differential expression between species, being evenly divided between C. remanei and C. latens having higher expression in one species or the other (nCrem = 3224, nClat = 3339), with the remaining 6865 genes showing conserved expression between the species (Figure 3.1A). 
```{r}
#Define coefficient vectors for each condition
cre = colMeans(mod_mat_wt[dds_wt$species == "Cre", ])
clat = colMeans(mod_mat_wt[dds_wt$species == "Clat", ])

# Perform contrast analysis between C. remanei and C. latens
crexclat = results(dds_wt, contrast = cre - clat, alpha = 0.05)  # Setting C. latens as the baseline

# Combine results with a new column indicating significance and direction of change
crexclat_res <- cbind(
  as.data.frame(crexclat), 
  as.data.frame(crexclat) %>% 
    mutate(crexclat = ifelse(padj > 0.05, 0, ifelse(log2FoldChange > 0, 1, -1))) %>% 
    dplyr::select(crexclat)
  )
# Remove rows with NA values
crexclat_res = na.omit(crexclat_res)

# Separate results into up/down regulated and conserved
crexclat_res1 = crexclat_res %>% filter(crexclat_res$crexclat != 0) # Upregulated and downregulated genes
crexclat_res2 = crexclat_res %>% filter(crexclat_res$crexclat == 0) #conserved expression genes

# Summarize the results
summary(crexclat)# Summary of results; upregulated in C. remanei with C. latens as the reference
nrow(crexclat_res) # Total Number of genes tested
table(crexclat_res$crexclat) # Count of species-biased and conserved genes

#Saving results
write.csv(crexclat_res, file = "../3-Output/crexclat_genes.csv")
```

###II BETWEEN TISSUES
FINDING GENES THAT ARE TISSUE-BIASED (INCLUDING THOSE THAT ARE DUE TO INTERATION TERM)

```{r}

#Define coefficient vectors for each condition
gonad = colMeans(mod_mat_wt[dds_wt$tissue == "G", ])
soma = colMeans(mod_mat_wt[dds_wt$tissue == "S", ])

# Perform contrast analysis comparing soma to gonad
gxs = results(dds_wt, contrast = soma - gonad, alpha = 0.05)  # Setting gonad as the baseline

# Combine results with a new column indicating significance and direction of change
gxs_res <- cbind(
  as.data.frame(gxs), 
  as.data.frame(gxs) %>% 
    mutate(gxs = ifelse(padj > 0.05, 0, ifelse(log2FoldChange > 0, 1, -1))) %>% 
    dplyr::select(gxs)
  )

# Remove rows with NA values
gxs_res = na.omit(gxs_res)

# Separate results into significant (up/downregualted) and non-significant (tissue neutral)
gxs_res1 = gxs_res %>% filter(gxs_res$gxs != 0) #tissue-biased genes
gxs_res2 = gxs_res %>% filter(gxs_res$gxs == 0) #tissue-neutral genes

# Summary of the results
nrow(gxs_res) #number of genes tested
summary(gxs)# Summary of results; upregulated in soma with gonad as the reference
table(gxs_res$gxs)

# Display the design formula used in DESeq2
design(dds_wt)

# Optionally, save results to a CSV file
write.csv(gxs_res, file = "../3-Output/gonadxsoma_genes.csv")
```

###III BETWEEN SEXES
FINDING GENES THAT ARE SEX-BIASED (INCLUDING THOSE THAT ARE DUE TO INTERATION TERM) 

```{r}
#Define coefficient vectors for each sex condition
male = colMeans(mod_mat_wt[dds_wt$sex == "M", ])
female = colMeans(mod_mat_wt[dds_wt$sex == "F", ])

# Perform contrast analysis comparing males to females
mxf = results(dds_wt, contrast = male - female, alpha = 0.05) # Setting female as the baseline

# Combine results with a new column indicating significance and direction of change
mxf_res <- cbind(
  as.data.frame(mxf), 
  as.data.frame(mxf) %>% 
    mutate(mxf = ifelse(padj > 0.05, 0, ifelse(log2FoldChange > 0, 1, -1))) %>% 
    dplyr::select(mxf)
  )

# Remove rows with NA values
mxf_res = na.omit(mxf_res)

# Separate results into sex-biased and sex-neutral genes
mxf_res1 = mxf_res %>% filter(mxf_res$mxf != 0)
mxf_res2 = mxf_res %>% filter(mxf_res$mxf == 0)

# Summary of the results
nrow(mxf_res) #number of genes tested
summary(mxf)# Summary of results; upregulated in males with females as the reference
table(mxf_res$mxf)

# Save results to a CSV file
write.csv(mxf_res, file = "../3-Output/mxf_genes.csv")
```

```{r}
#####COMBINING DATA
# Combine results from different analyses into a single data frame
all_total_biasgene_info = cbind(crexclat_res, gxs_res, mxf_res)  # Includes DEGs with interaction information

# Create a contingency table to summarize the relationships between the datasets
table(all_total_biasgene_info$crexclat, all_total_biasgene_info$gxs)  # First dataset as row names

table(all_total_biasgene_info$crexclat, all_total_biasgene_info$mxf) # First dataset as row names

```

###IV EFFECTS OF SPECIES & TISSUE DIFFERENCES
```{r}
# Perform contrast analysis 
spxts = results(dds_wt, contrast = list("speciesCre.tissueS"), alpha = 0.05) # Baseline is C. latens

# Combine results and create a new column indicating significance and direction of change
spxts_res <- cbind(
  as.data.frame(spxts), 
  as.data.frame(spxts) %>% 
    mutate(spxts = ifelse(padj > 0.05, 0, ifelse(log2FoldChange > 0, 1, -1))) %>% 
    dplyr::select(spxts)
  )

# Remove rows with NA values
spxts_res = na.omit(spxts_res)

# Separate significant and non-significant results
spxts_res1 = spxts_res %>% filter(spxts_res$spxts != 0 ) # Significant results
spxts_res2 = spxts_res %>% filter(spxts_res$spxts == 0 ) # 

# Summarize the results
summary(spxts)  # Summary of the analysis results
nrow(spxts_res)  # Number of genes
table(spxts_res$spxts)
```

###V EFFECTS OF SPECIES & SEX DIFFERENCES

```{r}
#Perform contrast analysis for species and sex differences
spxsex = results(dds_wt, contrast = list("speciesCre.sexM"), alpha = 0.05) # Baseline is C. latens

# Combine results and add a column indicating significance and direction of change
spxsex_res <- cbind(
  as.data.frame(spxsex), 
  as.data.frame(spxsex) %>% 
    mutate(spxsex = ifelse(padj > 0.05, 0, ifelse(log2FoldChange > 0, 1, -1))) %>% 
    dplyr::select(spxsex)
  )

# Remove rows with NA values
spxsex_res = na.omit(spxsex_res)

# Separate significant and non-significant results
spxsex_res1 = spxsex_res %>% filter(spxsex_res$spxsex != 0 ) # Significant results
spxsex_res2 = spxsex_res %>% filter(spxsex_res$spxsex == 0 ) # Non-Significant results

# Summarize the results
summary(spxsex)  # Summary of the analysis results
nrow(spxsex_res)  # Number of genes
table(spxsex_res$spxsex)
```



###VI EFFECTS OF TISSUE & SEX DIFFERENCES

```{r}
# Perform contrast analysis 
tsxsex = results(dds_wt, contrast = list("tissueS.sexM"), alpha = 0.05) #Baseline C. latens

#Combine results and add a column indicating significance and direction of change
tsxsex_res <- cbind(
  as.data.frame(tsxsex), 
  as.data.frame(tsxsex) %>% 
    mutate(tsxsex = ifelse(padj > 0.05, 0, ifelse(log2FoldChange > 0, 1, -1))) %>% 
    dplyr::select(tsxsex)
  )

tsxsex_res = na.omit(tsxsex_res)
tsxsex_res1 = tsxsex_res %>% filter(tsxsex_res$tsxsex != 0 )
tsxsex_res2 = tsxsex_res %>% filter(tsxsex_res$tsxsex == 0 )


#Summary of results
summary(tsxsex)
nrow(tsxsex_res)
table(tsxsex_res$tsxsex)
```


#3
##A
###I



