---
title: "4.Figures_WGCNA_WT"
author: "Athmaja Viswanath"
date: "2024-09-25"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
Figure labeling
A1_I_a - Section "1", subsection "A"/Figure A, sub-subsection "I"/figure I, first figure "a"

C3_I_a - Section "3", subsection "C", sub-subsection "I", first figure "a"
#0. Loading required packages
```{r}
library(WGCNA)
library(DESeq2)
library(tidyverse)
library(ggplot2)
library(ggstats)
library(CorLevelPlot)
library(gridExtra)
library(VennDiagram)
library(grid)
library(lme4)
library(dplyr)
library(ggsignif)
library(patchwork)
library(scales)

#`%notin%` <- Negate(`%in%`)
options(scipen = 999)  # Increase penalty to avoid scientific notation

```

```{r}
# module name information
module_number <- data.frame(Module_Colour = c("MEgrey", "MEmagenta", "MEbrown", "MEgreenyellow", "MEblack", "MEturquoise",
                                                  "MEgreen", "MEpink", "MEred", "MEblue", "MEyellow", "MEpurple"),
                                mcolour = c("grey", "magenta", "brown", "greenyellow", "black", "turquoise",
                                                  "green", "pink", "red", "blue", "yellow", "purple"),
                                module_number = c("M0", "M1", "M2", "M3", "M4", "M5", 
                                                  "M6", "M7", "M8", "M9", "M10", "M11"))
```

Figure labeling
A1_I_a - Section "1", subsection "A", sub-subsection "I", first figure "a"

C3_I_a - Section "3", subsection "C", sub-subsection "I", first figure "a"

#1. Plotting dendrogram 
this needs the bwnet (network constructed in 3.WGCNA_WT.Rmd), this code is also present after network creation

```{r}
# Plot the dendrogram with color assignments for both unmerged and merged modules
# plotDendroAndColors(bwnet$dendrograms[[1]], 
#                     cbind(bwnet$unmergedColors, bwnet$colors), 
#                     c("unmerged", "merged"), 
#                     dendroLabels = FALSE,
#                     oddGuide = TRUE,
#                     hang = 0.03, 
#                     guideHang = 0.05)
# 
# 
# dendo1 = plotDendroAndColors(bwnet$dendrograms[[1]], 
#                              cbind(bwnet$colors),
#                              c("merged"),
#                              dendroLabels = FALSE,
#                              oddGuide = TRUE, 
#                              hang = 0.03,
#                              guideHang = 0.05)
```

#2. Eigengene dendrogram

```{r}

module_eigengenes = read.csv("../3-Output/WGCNA/module eigengenes values_WT.csv")
row.names(module_eigengenes) = module_eigengenes$X
module_eigengenes = module_eigengenes[, -1]

#Eigengene Dendrograms 
eingen_dendo = plotEigengeneNetworks(module_eigengenes, "Eigengene dendrogram", marDendro = c(0,4,2,0),
                      plotHeatmaps = FALSE, excludeGrey = FALSE)

#Adjacency Matrix
eingen_heatmap = plotEigengeneNetworks(module_eigengenes, "Eigengene adjacency heatmap", marHeatmap = c(3,4,2,2),
                      plotDendrograms = FALSE, xLabelsAngle = 90, colorLabels = FALSE)

```

#3. Visualize the number of genes in each module

```{r}
###loading file which contains genes and their corresponding modules (obtained from above)
bwnet_df = read.table("../3-Output/WGCNA/WGCNA_modulegenes.txt", fill = T)


# Count the number of genes in each module
modules_counts = as.data.frame(table(bwnet_df$V2))

# Assuming modules_counts is your data frame
modules_counts$Module_num <- module_number$module_number[match(modules_counts$Var1, module_number$mcolour)]


# Create a bar plot of gene counts per module
A_3 = ggplot(modules_counts, aes(x=Module_num, y = Freq, fill = Module_num))+
  geom_bar(stat = "identity") +
  xlab(" ")+ 
  ylab("Number of genes")+
  theme(plot.title = element_text(size = 26, face = "bold", hjust = 0.5, vjust = 0.5))+
  coord_cartesian(ylim = c(0, 3500))+
  #scale_fill_manual(values = c( "#afced0", "#4b7d81", "#696967", '#DCDDDF'))+
  theme(axis.title.y = element_text(size = 14, hjust = 0.5))+
  theme(axis.text.x = element_text(size = 10,face = "bold", colour = "black"))+
  theme(axis.text.y = element_text(size = 10))+
  theme(axis.title.x = element_text(size = 14))+
  theme(axis.title.y = element_text(size = 20, hjust = 0.5))+
  theme(axis.text.y = element_text(color="black", size=15))+
  guides(fill=guide_legend(title=" "))+
  # scale_colour_brewer(palette = 2)+
  #scale_x_discrete(limits = c("Conserved","Differentially expressed", "C. nigoni dominant", "Ambiguous"))+ ##reordering character x-axis
  expand_limits(y=0)+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)), breaks=seq(0,3500,500))+
  ggtitle("WGCNA Modules")


# Save the plot
#ggsave("../3-Output/2.1WGCNA_Modules_Plot.pdf", plot, width = 10, height = 6, dpi = 300)
```

#4. Visualize chromosomal distribution of modules
##A. Getting and plotting wgcna data across chromosomes
```{r}
#Load file with orthologous gene names and their corresponding chromosomes
orthologs_chr = read.table(file = "../3-Output/orthologs_chr.txt", header = TRUE)

#Read module genes information 
WGCNA_modulegenes = read.table(file = "../3-Output/WGCNA/WGCNA_modulegenes.txt")
#changing rownames to gene names (first column and deleting the first column)
rownames(WGCNA_modulegenes) = WGCNA_modulegenes$V1 
#WGCNA_modulegenes = WGCNA_modulegenes[, -1]
head(WGCNA_modulegenes)


# Create a dataframe with non-NA entries based on orthologs_chr
WGCNA_modulegenes_a <- WGCNA_modulegenes[orthologs_chr$C..remanei.Gene.name_C.latens.Gene.name, ] %>%
  na.omit() 
  #rownames_to_column()
head(WGCNA_modulegenes_a)
# Update rownames in orthologs_chr
rownames(orthologs_chr) <- orthologs_chr$C..remanei.Gene.name_C.latens.Gene.name

# Combine dataframes for genes present in both datasets - modules and chromosomal info
WGCNA_modulegenes_chr <- na.omit(orthologs_chr[WGCNA_modulegenes_a$V1, ]) %>%
  cbind(WGCNA_modulegenes_a) %>%
  select(-V1) #11804 genes

WGCNA_modulegenes_chr <- WGCNA_modulegenes_chr %>% 
  filter(chromosome == "I" | chromosome == "II" | chromosome == "III" | chromosome == "IV" | chromosome == "V" | chromosome == "X") #11795 genes 

#Adding Autosome vs X column
WGCNA_modulegenes_chr = WGCNA_modulegenes_chr %>%
  mutate(A_X = ifelse(chromosome %in% c("I", "II", "III", "IV", "V"), "Autosome", "X"))

# adding module numbers 
WGCNA_modulegenes_chr$Module_num <- module_number$module_number[match(WGCNA_modulegenes_chr$V2, module_number$mcolour)]

WGCNA_modulegenes_chr_freq <- as.data.frame(table(WGCNA_modulegenes_chr$chromosome, WGCNA_modulegenes_chr$V2))
WGCNA_modulegenes_A_X_freq <- as.data.frame(table(WGCNA_modulegenes_chr$A_X, WGCNA_modulegenes_chr$V2))

View(WGCNA_modulegenes_chr_freq)
View(WGCNA_modulegenes_A_X_freq)

# Assigning module numbers
WGCNA_modulegenes_chr_freq$Module_num <- module_number$module_number[match(WGCNA_modulegenes_chr_freq$Var2, module_number$mcolour)]

WGCNA_modulegenes_A_X_freq$Module_num <- module_number$module_number[match(WGCNA_modulegenes_A_X_freq$Var2, module_number$mcolour)]


# Specify the desired order of Module_num
module_order <- c("M0", "M1", "M2", "M3", "M4", "M5", "M6", "M7", "M8", "M9", "M10", "M11")

##Plotting bar plot across all chromosomes

A_4 = ggplot(WGCNA_modulegenes_chr, aes(x = WGCNA_modulegenes_chr$chromosome, fill = factor(Module_num, levels = module_order))) + 
  geom_bar() +
  labs(x = "", y = "Number of genes", title = "Distribution of WGCNA modules") +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 20, hjust = 0.5),
    axis.text.x = element_text(size = 26, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 15, color = "black"),
    axis.title.x = element_text(size = 14)) +
  expand_limits(y = 0) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)), breaks = seq(0, 6500, 500)) +
  guides(fill = guide_legend(title = " ")) +
  labs(title = "Module Genes across Chromosomes", y = "Number of genes", x = "Chromosome") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


table(WGCNA_modulegenes_chr$chromosome, WGCNA_modulegenes_chr$Module_num)

# Proportion of modules on all chromosomes

B_4 = ggplot(WGCNA_modulegenes_chr_freq) +
  aes(x = Var1, fill = factor(Module_num, levels = module_order), weight = Freq, by = Var1) +
  geom_bar(position = "fill") +
  geom_text(stat = "prop", position = position_fill(.5))+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)), breaks=seq(0,1,0.25)) +
  labs(title = "Module Proportion across Chromosomes", y = "Number of genes", x = "Chromosome") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


##Plotting bar plot across Autosomes vs X-chromsome

C_4 = ggplot(WGCNA_modulegenes_chr, aes(x = WGCNA_modulegenes_chr$A_X, fill = factor(Module_num, levels = module_order))) + 
  geom_bar() +
  labs(x = "", y = "Number of genes", title = "Distribution of WGCNA modules") +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 20, hjust = 0.5),
    axis.text.x = element_text(size = 26, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 15, color = "black"),
    axis.title.x = element_text(size = 14)) +
  expand_limits(y = 0) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)), breaks = seq(0, 10000, 500)) +
  guides(fill = guide_legend(title = " ")) +
  labs(title = "Module Proportion across autosome and X Chromosomes", y = "Number of genes", x = "Chromosome") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  guides(fill = "none")  # Remove legend

table(WGCNA_modulegenes_chr$chromosome, WGCNA_modulegenes_chr$Module_num)

# Proportion of modules on all chromosomes

D_4 = ggplot(WGCNA_modulegenes_A_X_freq) +
  aes(x = Var1, fill = factor(Module_num, levels = module_order), weight = Freq, by = Var1) +
  geom_bar(position = "fill") +
  geom_text(stat = "prop", position = position_fill(.5))+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)), breaks=seq(0,1,0.25)) +
  labs(title = "Module Proportion across autosome and X Chromosomes", y = "Proportion", x = "Chromosome") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  guides(fill = "none")  # Remove legend


```

##B Performing Enrichment analysis 
###I. Hypergeometric test
```{r}
# Dataframe with gene names, modules, and chromosome information
gene_info <- data.frame(
  gene = c("gene1", "gene2", "gene3", "gene4", "gene5", 
           "gene6", "gene7", "gene8", "gene9"),
  module = c("Module1", "Module1", "Module1", "Module2", "Module2", 
             "Module3", "Module3", "Module3", "Module3"),
  chromosome = c("a", "a", "A", "a", "A", "A", "a", "A", "a")
)
#actual data
names(WGCNA_modulegenes_chr)

# Function to check enrichment for all modules in both chromosomes
check_enrichment_all <- function(WGCNA_modulegenes_chr) {
  results <- data.frame(Module = character(), Chromosome = character(), P_Value = numeric())
  
  unique_modules <- unique(WGCNA_modulegenes_chr$Module_num)
  
  for (module_name in unique_modules) {
    for (chromosome in c("X", "Autosome")) {
      module_genes <- WGCNA_modulegenes_chr$C..remanei.Gene.name_C.latens.Gene.name[WGCNA_modulegenes_chr$Module_num == module_name]
      
      # Count total genes in the specified chromosome
      total_genes_chromosome <- sum(WGCNA_modulegenes_chr$A_X == chromosome)
      total_genes_other <- nrow(WGCNA_modulegenes_chr) - total_genes_chromosome
      
      # Number of genes in the module that are in the specified chromosome
      genes_in_module <- sum(module_genes %in% WGCNA_modulegenes_chr$C..remanei.Gene.name_C.latens.Gene.name[WGCNA_modulegenes_chr$A_X == chromosome])
      
      # Hypergeometric test
      p_value <- phyper(genes_in_module - 1, total_genes_chromosome, total_genes_other, length(module_genes), lower.tail = FALSE)
      
      results <- rbind(results, data.frame(Module = module_name, Chromosome = chromosome, P_Value = p_value))
    }
  }
  
  return(results)
}


# Check enrichment for all modules
enrichment_results <- check_enrichment_all(WGCNA_modulegenes_chr)

# Plot the results using ggplot2 and scales
# library(scales)  
E_4 = ggplot(enrichment_results, aes(x = Module, y = P_Value, fill = Chromosome)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_hline(yintercept = 0.05, linetype = "dashed", color = "red") +  # Significance line at p = 0.05
  scale_y_continuous(trans = 'log10', breaks = scales::trans_breaks("log10", function(x) 10^x), labels = scales::trans_format("log10", math_format(10^.x))) +
  labs(title = "Module Enrichment in Chromosomes", y = "P-Value (log scale)", x = "Module") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))



```

###II Odds Ratio enrichmetn plots across Autosomes and X 
```{r}
WGCNA_modulegenes_chr_a_x = WGCNA_modulegenes_chr %>% group_by(A_X, Module_num) %>% dplyr::count()
df <- data.frame(A_X = c("X"), Module_num = c("M1"), n = c(1)) #prxy instead of 0 count for M1 module for X- chromosome

WGCNA_modulegenes_chr_a_x = rbind(WGCNA_modulegenes_chr_a_x, df)
WGCNA_modulegenes_chr_a_x$enrichement = as.numeric(enrichment(matrix(WGCNA_modulegenes_chr_a_x$n, nrow = 12), odds.ratio = T)) #nrow is basically total rows/2 (number of rowas for each category)
WGCNA_modulegenes_chr_a_x$pval = as.numeric(enrichment(matrix(WGCNA_modulegenes_chr_a_x$n, nrow = 12), odds.ratio = F))
WGCNA_modulegenes_chr_a_x$sig = ""
WGCNA_modulegenes_chr_a_x$sig[ WGCNA_modulegenes_chr_a_x$pval < 0.05 & abs(log2(WGCNA_modulegenes_chr_a_x$enrichement)) > 0.5 ] = "*"
WGCNA_modulegenes_chr_a_x$Sample = "WGCNA"

# Specify the desired order of Module_num
module_order_rev <- c("M11", "M10", "M9", "M8", "M7", "M6", "M5", "M4", "M3", "M2", "M1", "M0")

WGCNA_modulegenes_chr_a_x_2 = as.data.frame(table(WGCNA_modulegenes_chr$A_X, WGCNA_modulegenes_chr$Module_num))
F_4 = ggplot(WGCNA_modulegenes_chr_a_x %>% filter(A_X == "X"), aes(x=log2(enrichement), y=factor(Module_num, levels = module_order_rev))) +
  #geom_rect(data=rectX, aes(xmin=x1, xmax=x2, ymin=y1, ymax=y2), fill="grey90", alpha=0.5, inherit.aes=F) +
  geom_vline(xintercept=0, size=1) +
  geom_bar(aes(fill=Module_num), stat="identity", position=position_dodge(width=0.9), alpha=0.7, show.legend=F)+
  annotate(geom="label", label="autosomes", y=13, x=-.75, fill="black", color="white") +
  annotate(geom="label", label="X", y=13, x=.75, fill="black", color="white") +
  #facet_wrap(~Sample)+
  labs(y="", x=expression(paste("enrichment (log"[2], " odds ratio)"))) +
  background_grid(major="x", size.major = 0.2, colour.major = "grey75") +
  geom_text(aes(x=log2(enrichement)/2, label=sig), size=5, show.legend=F, color="white") +
  theme(legend.text=element_text(hjust=0),
        strip.background = element_rect(fill="grey90", color=NA))+
  scale_x_continuous(limits = c(-10, 10),
                     breaks = seq(-10, 10, by = 2),
                     labels = c("-10", "-8", "-6", "-4", "-2", "0", "2", "4", "6", "8", "10"))
```

##C. Organizing chromosomal plots
```{r}
# Arrange plots in a single panel
chromosomal_modules <- A_4+ B_4 + C_4 + D_4 +E_4 +F_4 + plot_layout(ncol = 2)


# Saving as pdf
# ggsave("../3-Output/WGCNA/Modules across chromosomes.pdf", plot = chromosomal_modules, width = 34, height = 15)
# 
# # # #Saving as png
# ggsave("../3-Output/WGCNA/Modules across chromosomes.png", plot = chromosomal_modules, width = 34, height = 15)

```

#5. Visualize Eigengene Plots for each module
```{r}
#Make a text file with average module eigen gene values across replicates for each module

# Load average counts data
average_counts <- read.table("../3-Output/WGCNA/Average_MEvalues.txt", sep = "\t", header = TRUE, comment.char = "#")

# Add metadata columns
average_counts$Species <- rep(c("Cre", "Clat"), each = 4)
average_counts$Sex <- rep(c("F", "F", "M", "M"), times = 2)
average_counts$Tissue <- rep(c("G", "S"), times = 4)
average_counts$S_Sp <- paste(average_counts$Species, average_counts$Sex, sep = "_")

# Display structure of the data
str(average_counts)

######################
# Plot for each module
######################

# Turquoise
average_counts %>% 
  ggplot(aes(Tissue, MEturquoise, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  stat_smooth(method = "lm") +
  ggtitle("M5 N = 3156")


# BLUE
average_counts %>% 
  ggplot(aes(Tissue, MEblue, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  stat_smooth(method = "lm") +
  ggtitle("M9 N = 2761")

# GREENYELLOW
average_counts %>% 
  ggplot(aes(Tissue, MEgreenyellow, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M3 N = 133")

# PURPLE
average_counts %>% 
  ggplot(aes(Tissue, MEpurple, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M11 N = 164")

# RED
average_counts %>% 
  ggplot(aes(Tissue, MEred, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M8 N = 652")

# BLACK
average_counts %>% 
  ggplot(aes(Tissue, MEblack, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M4 N = 583")

# GREEN
average_counts %>%
  ggplot(aes(Tissue, MEgreen, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M6 N = 822")


# YELLOW
average_counts %>% 
  ggplot(aes(Tissue, MEyellow, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M10 N = 979")

# PINK
average_counts %>% 
  ggplot(aes(Tissue, MEpink, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M7 N = 398")

# BROWN
average_counts %>% 
  ggplot(aes(Tissue, MEbrown, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M2 N = 1770")

# MAGENTA
average_counts %>% 
  ggplot(aes(Tissue, MEmagenta, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M1 N = 274")


# GREY
average_counts %>% 
  ggplot(aes(Tissue, MEgrey, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M0 N = 112")

```


#6. Visualize module trait association as a heatmap
##A.Preparing, data, coldata and binarizing data (from 3.WGCNA_WT)
```{r}
#Data preparation 
##A.Separating orthologous gene names so that they can be used to filter readcounts for each species.
# Load ortholog data from a tab-separated file, ignoring comments
orthologs = read.table("../1-Input/new_1to1_orthologgenelist.txt", sep="\t", header=TRUE, comment.char="#")

# Extract the relevant columns into separate variables
cre_ortho = orthologs[, 1]            # First column: cre_ortho, Cre names
clat_ortho = orthologs[, 2]           # Second column: clat_ortho, Clat names
cumulative_ortho = orthologs[, 3]     # Third column: cumulative_ortho, joint names

#########
#C.remanei
#########
# Load counts for C. remanei
cre_counts = read.table("../1-Input/C.remanei_cumulative.txt", sep="\t", header=TRUE, row.names=1, comment.char="#")
# Look at the structure of data
head(cre_counts)
colnames(cre_counts)
nrow(cre_counts) #20091
# Filter to include only orthologs and exclude Whole animal samples
cre_counts_ortho = cre_counts[cre_ortho, -c(13:15)] 
# Update row names to match cumulative orthologs
rownames(cre_counts_ortho) = cumulative_ortho 
# Check the number of rows and column names
nrow(cre_counts_ortho)
colnames(cre_counts_ortho)

#########
#C.latens
#########
# Load counts for C. latens
clat_counts = read.table("../1-Input/C.latens_cumulative.txt", sep="\t", header=TRUE, row.names=1, comment.char="#")
# Look at the structure of data
head(clat_counts)
colnames(clat_counts)
nrow(clat_counts) #23073
# Filter to include only orthologs and exclude Whole animal samples
clat_counts_ortho = clat_counts[clat_ortho, -c(13:15)]
# Update row names to match cumulative orthologs
rownames(clat_counts_ortho) = cumulative_ortho
# Check the number of rows and column names for C. latens
nrow(clat_counts_ortho)
names(clat_counts_ortho)


# Combine ortholog counts from C. remanei and C. latens
wt_counts_orthologs = cbind(cre_counts_ortho, clat_counts_ortho)

# Display current column names
colnames(wt_counts_orthologs)

# Clean up column names by removing the "X" prefix
colnames(wt_counts_orthologs) = gsub("X", "", colnames(wt_counts_orthologs))
colnames(wt_counts_orthologs)
```
```{r}
# Prepare the coldata data frame
sample_name = colnames(wt_counts_orthologs)  # Get sample names
tissue = substr(sample_name, 3, 3)            # Extract tissue type from sample names
sex = substr(sample_name, 2, 2)               # Extract sex from sample names
species = c(rep("Cre", 12), rep("Clat", 12))  # Assign species labels
batch = rep(c(1, 2, 3),8)                # Assign batch numbers

# Create the coldata data frame
coldata = data.frame(sample_name, species, sex, tissue, batch)

# Convert columns to factors for analysis
coldata$species = factor(coldata$species)
coldata$tissue = factor(coldata$tissue)
coldata$sex = factor(coldata$sex)
coldata$batch = factor(coldata$batch)

# Display the coldata data frame
coldata  # Note: WM samples are not included

# Prepare column data by setting the first column as row names
coldata2 = coldata[, -1]  # Remove the first column
rownames(coldata2) = coldata[, 1]  # Set the first column as row names

# Check if all row names in coldata2 match the column names of counts_orthologs
all(rownames(coldata2) %in% colnames(wt_counts_orthologs))
all(rownames(coldata2) == colnames(wt_counts_orthologs))
```

```{r}
# Binarizing trait data

###1.Sexes - has three categories - M, F, W
#coldata2$sex = factor(coldata2$sex, levels = c("M", "F", "W")) #M will nto showup after binarised below

sex.out = binarizeCategoricalColumns(coldata2$sex,
                                     includePairwise = TRUE,
                                     includeLevelVsAll = FALSE)

row.names(sex.out) = row.names(coldata2) #need to change rownames when binarizing categorical variables


###2. Species
# species = coldata2 %>%
#   mutate(species_binary = ifelse(grepl("Cre", species), 1, 0)) %>% #merges M & W into 0s
#   select(5)

species.factor <- factor(coldata2$species,
                         levels = c('Clat','Cre','H1','H2')) #changing the reference level
species = binarizeCategoricalVariable(coldata2$species,
                                      includePairwise = TRUE,
                                      includeLevelVsAll = FALSE) ##o is the reference category
row.names(species) = row.names(coldata2) #need to change rownames when binarizing categorical variables

###3. TISSUES

tissues = binarizeCategoricalVariable(coldata2$tissue,
                                      includePairwise = TRUE,
                                      includeLevelVsAll = FALSE) ##o is the reference category
row.names(tissues) = row.names(coldata2) #need to change rownames when binarizing categorical variables


##Combining binarized data 

traits = cbind(sex.out, species, tissues)


```

```{r}
heatmap_data = merge(module_eigengenes, traits, by = "row.names")
heatmap_data = heatmap_data %>% 
  column_to_rownames(var = "Row.names")

# Create a named vector for mapping
module_map <- c("MEmagenta" = "M1", "MEbrown" = "M2", "MEgreenyellow" = "M3", 
                "MEblack" = "M4", "MEturquoise" = "M5", "MEgreen" = "M6", 
                "MEpink" = "M7", "MEred" = "M8", "MEblue" = "M9", 
                "MEyellow" = "M10", "MEpurple" = "M11", "MEgrey" = "M0", 
                "data.M.vs.F" = "Males vs Females", "Cre.vs.Clat" = "Cre.vs.Clat", "S.vs.G" = "Soma vs Gonad")

# Rename columns using the module_map
colnames(heatmap_data) <- module_map[colnames(heatmap_data)]

colnames(heatmap_data)
a6 = CorLevelPlot(heatmap_data,
             x = names(heatmap_data)[13], 
             y = names(heatmap_data)[1:12],
             col = c("red", "pink","white", "pink", "red"),
             main = "SEX")

b6 = CorLevelPlot(heatmap_data,
             x = names(heatmap_data)[14], 
             y = names(heatmap_data)[1:12],
             col = c("#ffeb3b","#fff280", "white", "#fff280", "#ffeb3b"),
             main = "SPECIES")

c6 = CorLevelPlot(heatmap_data,
             x = names(heatmap_data)[15], 
             y = names(heatmap_data)[1:12],
             col = c("#004b8d","skyblue", "white", "skyblue", "#004b8d" ),
             main = "TISSUES")

d6 = CorLevelPlot(heatmap_data,
             x = names(heatmap_data)[13:15], 
             y = names(heatmap_data)[1:12],
             col = c("blue1", "skyblue", "white", "pink", "red" ), 
             main = "All Conditions")

###Modules in red/blue are significantly associated with one of the sex/species over the other
```


#7. Gene significance vs. module membership plots - intramodular hub genes
##A. Species-biased modules
###I. Yellow (M10) module
```{r}
# Map genes to modules
module.gene.mapping = bwnet_df %>% column_to_rownames(var = "V1")

# YELLOW MODULE or M10

# Get genes in the yellow module
yellow_module = module.gene.mapping %>% 
  filter(V2 == "yellow") %>% 
  rownames()

# Get module membership values for genes in yellow module (M10)
# Note: module.mem.measure is transposed, so rows represent module names and columns represent gene names

mem_yellow = as.data.frame(t(module.mem.measure[2,yellow_module]))
gs_yellow = as.data.frame(abs(gene.sig.cor[yellow_module, 1]))

# Combine membership and gene significance data
yelow_gs_meme = cbind(mem_yellow, gs_yellow)

## Create a new column to categorize genes as "hub" or "non-hub"
yelow_gs_meme$category = with(yelow_gs_meme, ifelse(MEyellow>0.9 & `abs(gene.sig.cor[yellow_module, 1])`>0.9, "hub genes", "non-hub"))


#####Gene significance vs module membership plot
A7_I_a = ggplot(yelow_gs_meme, aes(x=yelow_gs_meme$MEyellow, y=yelow_gs_meme$`abs(gene.sig.cor[yellow_module, 1])`, colour = yelow_gs_meme$category)) +
  geom_point(alpha = 0.6, show.legend = FALSE)+
  coord_cartesian(ylim = c(0, 1))+
  coord_cartesian(xlim = c(0, 1))+
  geom_hline(yintercept = 0.9, colour = "red")+
  geom_vline(xintercept = 0.9, colour = "red")+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))+
  scale_x_continuous(expand = expansion(mult = c(0, 0.1)))+
  labs(title = "Intramodular hub genes for M10 (Yellow, species-biased)",
       x = "Module Membership",
       y = "Gene Significance") +
  theme(
    plot.title = element_text(size = 22, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
  )
table(yelow_gs_meme$category)
```


###II. Black (M4) module
```{r}
# Map genes to modules
module.gene.mapping = bwnet_df %>% column_to_rownames(var = "V1")

# BLACK MODULE

# Get genes in the black module
black_module = module.gene.mapping %>% 
  filter(V2 == "black") %>% 
  rownames()

# Get module membership values for genes in yellow module (M10)
# Note: module.mem.measure is transposed, so rows represent module names and columns represent gene names

mem_black = as.data.frame(t(module.mem.measure[8,black_module]))
gs_black = as.data.frame(abs(gene.sig.cor[black_module, 1]))

# Combine membership and gene significance data
black_gs_meme = cbind(mem_black, gs_black)

## Create a new column to categorize genes as "hub" or "non-hub"
black_gs_meme$category = with(black_gs_meme, ifelse(MEblack>0.9 & `abs(gene.sig.cor[black_module, 1])`>0.9, "hub genes", "non-hub"))

table(black_gs_meme$category)

#####Gene significance vs module membership plot
A7_II_a = ggplot(black_gs_meme, aes(x=black_gs_meme$MEblack, y=black_gs_meme$`abs(gene.sig.cor[black_module, 1])`, colour = black_gs_meme$category)) +
  geom_point(alpha = 0.6, show.legend = FALSE)+
  coord_cartesian(ylim = c(0, 1))+
  coord_cartesian(xlim = c(0, 1))+
  geom_hline(yintercept = 0.9, colour = "red")+
  geom_vline(xintercept = 0.9, colour = "red")+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))+
  scale_x_continuous(expand = expansion(mult = c(0, 0.1)))+
  labs(title = "Intramodular hub genes for M4 (Black, Species-biased)",
       x = "Module Membership",
       y = "Gene Significance") +
  theme(
    plot.title = element_text(size = 22, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
  )

```

##B. Tissue-biased modules
###I. Magenta (M1) module
```{r, warning=FALSE}
# Map genes to modules
module.gene.mapping = bwnet_df %>% column_to_rownames(var = "V1")

# Magenta (M1) module

# Get genes in the magenta module
magenta_module = module.gene.mapping %>% 
  filter(V2 == "magenta") %>% 
  rownames()

# Get module membership values for genes in magenta module (M1)
# Note: module.mem.measure is transposed, so rows represent module names and columns represent gene names

mem_magenta = as.data.frame(t(module.mem.measure[11,magenta_module]))
gs_magenta = as.data.frame(abs(gene.sig.cor.ts[magenta_module, 1]))

# Combine membership and gene significance data
magenta_gs_meme = cbind(mem_magenta, gs_magenta)

## Create a new column to categorize genes as "hub" or "non-hub"
magenta_gs_meme$category = with(magenta_gs_meme, ifelse(MEmagenta>0.9 & `abs(gene.sig.cor.ts[magenta_module, 1])`>0.9, "hub genes", "non-hub"))


#####Gene significance vs module membership plot
B7_I_a = ggplot(magenta_gs_meme, aes(x=magenta_gs_meme$MEmagenta, y=magenta_gs_meme$`abs(gene.sig.cor.ts[magenta_module, 1])`, colour = magenta_gs_meme$category)) +
  geom_point(alpha = 0.6, show.legend = FALSE)+
  coord_cartesian(ylim = c(0, 1))+
  coord_cartesian(xlim = c(0, 1))+
  geom_hline(yintercept = 0.9, colour = "red")+
  geom_vline(xintercept = 0.9, colour = "red")+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))+
  scale_x_continuous(expand = expansion(mult = c(0, 0.1)))+
  labs(title = "Intramodular hub genes for M1 (Magenta, Tissue-biased)",
       x = "Module Membership",
       y = "Gene Significance") +
  theme(
    plot.title = element_text(size = 22, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
  )

```

###II. Blue (M9) module
```{r}
# Blue (M9) module

# Get genes in the blue module
blue_module = module.gene.mapping %>% 
  filter(V2 == "blue") %>% 
  rownames()

# Get module membership values for genes in blue module (M9)
# Note: module.mem.measure is transposed, so rows represent module names and columns represent gene names

mem_blue = as.data.frame(t(module.mem.measure[3,blue_module]))
gs_blue = as.data.frame(abs(gene.sig.cor.ts[blue_module, 1]))

# Combine membership and gene significance data
blue_gs_meme = cbind(mem_blue, gs_blue)

## Create a new column to categorize genes as "hub" or "non-hub"
blue_gs_meme$category = with(blue_gs_meme, ifelse(MEblue>0.9 & `abs(gene.sig.cor.ts[blue_module, 1])`>0.9, "hub genes", "non-hub"))


#####Gene significance vs module membership plot
B7_II_a = ggplot(blue_gs_meme, aes(x=blue_gs_meme$MEblue, y=blue_gs_meme$`abs(gene.sig.cor.ts[blue_module, 1])`, colour = blue_gs_meme$category)) +
  geom_point(alpha = 0.6, show.legend = FALSE)+
  coord_cartesian(ylim = c(0, 1))+
  coord_cartesian(xlim = c(0, 1))+
  geom_hline(yintercept = 0.9, colour = "red")+
  geom_vline(xintercept = 0.9, colour = "red")+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))+
  scale_x_continuous(expand = expansion(mult = c(0, 0.1)))+
  labs(title = "Intramodular hub genes for M9 (Blue, Tissue-biased)",
       x = "Module Membership",
       y = "Gene Significance") +
  theme(
    plot.title = element_text(size = 22, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
  )

```

##C. Sex-biased modules
###I. Brown (M2) module
```{r}

# Get genes in the brown (M2) module
brown_module = module.gene.mapping %>% 
  filter(V2 == "brown") %>% 
  rownames()

# Get module membership values for genes in brown module (M10)
# Note: module.mem.measure is transposed, so rows represent module names and columns represent gene names

mem_brown = as.data.frame(t(module.mem.measure[10,brown_module]))
gs_brown = as.data.frame(abs(gene.sig.cor.sex[brown_module, 1]))

# Combine membership and gene significance data
brown_gs_meme = cbind(mem_brown, gs_brown)

## Create a new column to categorize genes as "hub" or "non-hub"
brown_gs_meme$category = with(brown_gs_meme, ifelse(MEbrown>0.9 & `abs(gene.sig.cor.sex[brown_module, 1])`>0.9, "hub genes", "non-hub"))


#####Gene significance vs module membership plot
C7_I_a = ggplot(brown_gs_meme, aes(x=brown_gs_meme$MEbrown, y=brown_gs_meme$`abs(gene.sig.cor.sex[brown_module, 1])`, colour = brown_gs_meme$category)) +
  geom_point(alpha = 0.6, show.legend = FALSE)+
  coord_cartesian(ylim = c(0, 1))+
  coord_cartesian(xlim = c(0, 1))+
  geom_hline(yintercept = 0.9, colour = "red")+
  geom_vline(xintercept = 0.9, colour = "red")+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))+
  scale_x_continuous(expand = expansion(mult = c(0, 0.1)))+
  labs(title = "Intramodular hub genes for M2 (Brown, Sex-biased)",
       x = "Module Membership",
       y = "Gene Significance") +
  theme(
    plot.title = element_text(size = 22, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
  )

```

###II. Pink (M7) module
```{r}
# Get genes in the pink (M7) module
pink_module = module.gene.mapping %>% 
  filter(V2 == "pink") %>% 
  rownames()

# Get module membership values for genes in pink module (M7)
# Note: module.mem.measure is transposed, so rows represent module names and columns represent gene names

mem_pink = as.data.frame(t(module.mem.measure[5,pink_module]))
gs_pink = as.data.frame(abs(gene.sig.cor.sex[pink_module, 1]))

# Combine membership and gene significance data
pink_gs_meme = cbind(mem_pink, gs_pink)

## Create a new column to categorize genes as "hub" or "non-hub"
pink_gs_meme$category = with(pink_gs_meme, ifelse(MEpink>0.9 & `abs(gene.sig.cor.sex[pink_module, 1])`>0.9, "hub genes", "non-hub"))


#####Gene significance vs module membership plot
C7_II_a = ggplot(pink_gs_meme, aes(x=pink_gs_meme$MEpink, y=pink_gs_meme$`abs(gene.sig.cor.sex[pink_module, 1])`, colour = pink_gs_meme$category)) +
  geom_point(alpha = 0.6, show.legend = FALSE)+
  coord_cartesian(ylim = c(0, 1))+
  coord_cartesian(xlim = c(0, 1))+
  geom_hline(yintercept = 0.9, colour = "red")+
  geom_vline(xintercept = 0.9, colour = "red")+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))+
  scale_x_continuous(expand = expansion(mult = c(0, 0.1)))+
  labs(title = "Intramodular hub genes for M7 (Pink, Sex-biased)",
       x = "Module Membership",
       y = "Gene Significance") +
  theme(
    plot.title = element_text(size = 22, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
  )

```

##D. Organizing hub gene plots
```{r, warning=FALSE}
# Arrange plots in a single panel
intramodular_hubgenes <- A7_I_a + A7_II_a + B7_I_a + B7_II_a + C7_I_a + C7_II_a + plot_layout(ncol = 2)
table(pink_gs_meme$category)
table(brown_gs_meme$category)
table(magenta_gs_meme$category)
table(blue_gs_meme$category)
table(yelow_gs_meme$category)
table(black_gs_meme$category)

# Saving as pdf
# ggsave("../3-Output/WGCNA/intramodular_hubgenes.pdf", plot = intramodular_hubgenes, width = 34, height = 15)
# 
# #Saving as png
# ggsave("../3-Output/WGCNA/intramodular_hubgenes.png", plot = intramodular_hubgenes, width = 34, height = 15)


# Getting number of hub and non-hub genes
# Create tables
## Species-biased
yellow_table <- table(yelow_gs_meme$category)
black_table <- table(black_gs_meme$category)
# Tissue-biased
magenta_table <- table(magenta_gs_meme$category)
blue_table <- table(blue_gs_meme$category)
# Sex-biased
pink_table <- table(pink_gs_meme$category)
brown_table <- table(brown_gs_meme$category)

# Convert tables to data frames
yellow_df <- as.data.frame(yellow_table)
black_df <- as.data.frame(black_table)
magenta_df <- as.data.frame(magenta_table)
blue_df <- as.data.frame(blue_table)
pink_df <- as.data.frame(pink_table)
brown_df <- as.data.frame(brown_table)

# Rename columns for clarity
colnames(yellow_df) <- c("category", "yellow_count")
colnames(black_df) <- c("category", "black_count")
colnames(magenta_df) <- c("category", "magenta_count")
colnames(blue_df) <- c("category", "blue_count")
colnames(pink_df) <- c("category", "pink_count")
colnames(brown_df) <- c("category", "brown_count")



# Merge all data frames into one
intramodular_hubgene_counts <- Reduce(function(x, y) merge(x, y, by = "category", all = TRUE), 
                     list(yellow_df, black_df, magenta_df, blue_df, pink_df, brown_df))

intramodular_hubgene_counts= as.data.frame(intramodular_hubgene_counts)

# Uncomment to write the results to csv file
# write.csv(intramodular_hubgene_counts, file = "../3-Output/WGCNA/intramodular_hubgene_counts.csv",
#             row.names = FALSE, col.names = FALSE, quote = FALSE)
```

#8. Venn Diagrams
##A. Visualize the overlap between species-biased hub genes (from WGCNA) and species-biased genes (from DESeq2)

```{r}
crexclat_res <- read.csv("../3-Output/DEGAnalysis/crexclat_genes.csv")

cre_biased = subset(crexclat_res, crexclat_res$crexclat == 1) %>% select("X")
clat_biased = subset(crexclat_res, crexclat_res$crexclat == -1) %>% select("X")

############
#YELLOW(M10)
############

##Overlap with Yellow (M10) genes
M10_hubgenes = subset(yelow_gs_meme, yelow_gs_meme$category=="hub genes") %>% row.names()
M10_allgenes  = yelow_gs_meme %>% row.names()
M10_clat_overlap = clat_biased %>% filter(X %in% M10_hubgenes) #as M10 has Clat-biased genes only

##Venndiagram
#NOTE: RUN THE VENNDIAGRAM CODE IN CONSOLE

# Create a Venn diagram without saving to a file
# temp <- venn.diagram(
#   list("Clat_biased" = clat_biased, 
#        "M10 hub genes" = M10_hubgenes, 
#        "M10 all genes" = M10_allgenes),
#   filename = NULL
# )
# 
# # Draw a triple Venn diagram with specified parameters
# temp = draw.triple.venn(979, 3339, 271, 952, 271, 271, 271, fileName = NULL)
# 
# 
# # Save the Venn diagram as a PDF
# pdf(file = "../3-Output/DEGvsHubgenes_M10.pdf")
# grid.draw(temp)
# dev.off()  # Close the PDF device

#NOTE: if it doesn't work in code chunk, try clearing devices (below) then re-reun in code chunk or console
# while (!is.null(dev.list())) {
#      dev.off()
#    }

# Check if any device is open before closing it
# if (dev.cur() > 1) {
#   dev.off()  # Close the PDF device if it is open
# } else {
#   message("No graphics device is currently open.")
# }

###########
#BLACK(M4)
###########

##Overlap with Black (M4) genes
M4_hubgenes = subset(black_gs_meme, black_gs_meme$category=="hub genes") %>% row.names()
M4_allgenes  = black_gs_meme %>% row.names()
M4_cre_overlap = cre_biased %>% filter(X %in% M4_hubgenes) #as M4 has Cre-biased genes only

##Venndiagram
#NOTE: RUN THE VENNDIAGRAM CODE IN CONSOLE

# Create a Venn diagram without saving to a file
# temp <- venn.diagram(
#   list("Cre_biased" = cre_biased$X, 
#        "M4 hub genes" = M4_hubgenes, 
#        "M4 all genes" = M4_allgenes),
#   filename = NULL
# )
# 
# # Save the Venn diagram as a PDF
# pdf(file = "../3-Output/DEGvsHubgenes_M4.pdf")
# grid.draw(temp)
# dev.off()  # Close the PDF device

```

##B. Visualize the overlap between tissue-biased hub genes (from WGCNA) and tissue-biased genes (from DESeq2)
```{r}
gxs_res <- read.csv("../3-Output/DEGAnalysis/gonadxsoma_genes.csv")

soma_biased = subset(gxs_res, gxs_res$gxs == 1) %>% select("X")
gonad_biased = subset(gxs_res, gxs_res$gxs == -1) %>% select("X")

############
#MAGENTA (M1)
############

##Overlap with MAGENTA (M1) genes
M1_hubgenes = subset(magenta_gs_meme, magenta_gs_meme$category=="hub genes") %>% row.names()
M1_allgenes  = magenta_gs_meme %>% row.names()
M1_gonad_overlap = gonad_biased %>% filter(X %in% M1_hubgenes) #as M1 has Gonad-biased genes only

##Venndiagram
#NOTE: RUN THE VENNDIAGRAM CODE IN CONSOLE

# Create a Venn diagram without saving to a file
temp <- venn.diagram(
  list("Gonad_biased" = gonad_biased,
       "M1 hub genes" = M1_hubgenes,
       "M1 all genes" = M1_allgenes),
  filename = NULL
)

# Save the Venn diagram as a PDF
pdf(file = "3-Output/DEGvsHubgenes_M1_magenta.pdf")
grid.draw(temp)
dev.off()  # Close the PDF device

#NOTE: if it doesn't work in code chunk, try clearing devices (below) then re-reun in code chunk or console
# while (!is.null(dev.list())) {
#      dev.off()
#    }

###########
#BLUE (M9)
###########

##Overlap with Black (M9) genes
M9_hubgenes = subset(blue_gs_meme, blue_gs_meme$category=="hub genes") %>% row.names()
M9_allgenes  = blue_gs_meme %>% row.names()
M9_soma_overlap = soma_biased %>% filter(X %in% M9_hubgenes) #as M9 has Soma-biased genes only

##Venndiagram
#NOTE: RUN THE VENNDIAGRAM CODE IN CONSOLE

# Create a Venn diagram without saving to a file
temp <- venn.diagram(
  list("Soma_biased" = soma_biased,
       "M9 hub genes" = M9_hubgenes,
       "M9 all genes" = M9_allgenes),
  filename = NULL
)

# Save the Venn diagram as a PDF
pdf(file = "3-Output/DEGvsHubgenes_M9_blue.pdf")
grid.draw(temp)
dev.off()  # Close the PDF device

```


##C. Visualize the overlap between sex-biased hub genes (from WGCNA) and sex-biased genes (from DESeq2)

```{r}
mxf_res <- read.csv("../3-Output/DEGAnalysis/mxf_genes.csv")

male_biased = subset(mxf_res, mxf_res$mxf == 1) %>% select("X")
female_biased = subset(mxf_res, mxf_res$mxf == -1) %>% select("X")

############
#BROWN (M2)
############

##Overlap with BROWN (M2) genes
M2_hubgenes = subset(brown_gs_meme, brown_gs_meme$category=="hub genes") %>% row.names()
M2_allgenes  = brown_gs_meme %>% row.names()
M2_male_overlap = male_biased %>% filter(X %in% M2_hubgenes) #as M2 has Male-biased genes only

##Venndiagram
#NOTE: RUN THE VENNDIAGRAM CODE IN CONSOLE

# Create a Venn diagram without saving to a file
temp <- venn.diagram(
  list("Male_biased" = male_biased,
       "M2 hub genes" = M2_hubgenes,
       "M2 all genes" = M2_allgenes),
  filename = NULL
)

# Save the Venn diagram as a PDF
pdf(file = "3-Output/DEGvsHubgenes_M2_brown.pdf")
grid.draw(temp)
dev.off()  # Close the PDF device

#NOTE: if it doesn't work in code chunk, try clearing devices (below) then re-reun in code chunk or console
# while (!is.null(dev.list())) {
#      dev.off()
#    }

###########
#PINK (M7)
###########

  ##Overlap with Pink (M9) genes
M7_hubgenes = subset(pink_gs_meme, pink_gs_meme$category=="hub genes") %>% row.names()
M7_allgenes  = pink_gs_meme %>% row.names()
M7_female_overlap = female_biased %>% filter(X %in% M7_hubgenes) #as M7 has Soma-biased genes only

##Venndiagram
#NOTE: RUN THE VENNDIAGRAM CODE IN CONSOLE

# Create a Venn diagram without saving to a file
temp <- venn.diagram(
  list("Female_biased" = female_biased,
       "M7 hub genes" = M7_hubgenes,
       "M7 all genes" = M7_allgenes),
  filename = NULL
)

# Save the Venn diagram as a PDF
pdf(file = "3-Output/DEGvsHubgenes_M7_pink.pdf")
grid.draw(temp)
dev.off()  # Close the PDF device

```

##D. Visualize the overlap between Male-biased modules (from WGCNA) and sex-biased genes (from DESeq2)

```{r}
mxf_res <- read.csv("../3-Output/DEGAnalysis/mxf_genes.csv")

male_biased = subset(mxf_res, mxf_res$mxf == 1) %>% select("X")
female_biased = subset(mxf_res, mxf_res$mxf == -1) %>% select("X")

############
#BROWN (M2)
############

##Overlap with BROWN (M2) genes
M2_allgenes  = brown_gs_meme %>% row.names()
M9_allgenes = blue_gs_meme

M2_male_overlap = male_biased %>% filter(X %in% M2_hubgenes) #as M2 has Male-biased genes only

##Venndiagram
#NOTE: RUN THE VENNDIAGRAM CODE IN CONSOLE

# Create a Venn diagram without saving to a file
temp <- venn.diagram(
  list("Male_biased" = male_biased,
       "M2 all genes" = M2_allgenes),
  filename = NULL
)

# Save the Venn diagram as a PDF
pdf(file = "../3-Output/DEGvsmodulegenes_M2_brown.pdf")
grid.draw(temp)
dev.off()  # Close the PDF device

#NOTE: if it doesn't work in code chunk, try clearing devices (below) then re-reun in code chunk or console
# while (!is.null(dev.list())) {
#      dev.off()
#    }


```

#9. Trends associated with effective number of codons (ENC) and dn/ds
##A. Preparing data to filter for genes with correct gene names along with enc, dn/ka and ds/ks values

```{r}
# Read data from text files
ka_ks_enc_data = read.table("../Specific specific gene confirmation/New annotation/FitMG94_dN_dS_ENC_corrected_dS_values_all_orthologs_new_annotations_remanei_latens.txt", header = T)

# Check the structure data
View(ka_ks_enc_data) #13793 genes
# View(remanei_gene_names) #15276 genes
# View(latens_gene_names) #18195 genes

#Calculate dn/ds_corrected using ds_corrected for ENC value

ka_ks_enc_data$dn.ds_corrected = ka_ks_enc_data$dN/ka_ks_enc_data$dS_corrected

ka_ks_enc_data$Remanei_Gene <- gsub("-R.*$", "", ka_ks_enc_data$Remanei_Gene)
ka_ks_enc_data$Latens_Gene <- gsub("-R.*$", "", ka_ks_enc_data$Latens_Gene)
orthologs = orthologs[order(orthologs$C..remanei.Gene.name), ]

#######################################################
# Filter ENC data for orthologous genes
head(orthologs)
ortho_enc_data <- ka_ks_enc_data %>% filter(ka_ks_enc_data$Remanei_Gene %in% orthologs$C..remanei.Gene.name) #13785 genes
ortholog_enc_kaks_data <- cbind(ortho_enc_data, orthologs$C..remanei.Gene.name_C.latens.Gene.name)
colnames(ortholog_enc_kaks_data)[9] <- "ortho_names"

# Save the data
# Uncomment to write the results to text files
# write.table(ortholog_enc_kaks_data, file = "../3-Output/ortholog_enc_kaks_data.txt",
#              row.names = FALSE, col.names = TRUE, quote = FALSE)
```

##B. Filter ENC data for diffferent modules
```{r}
#Using chromosomal and module information from section #4
str(WGCNA_modulegenes_chr)

# Use the previously made ka, ks, enc data
str(ortholog_enc_kaks_data)

WGCNA_modulegenes_chr = WGCNA_modulegenes_chr %>%
  arrange(WGCNA_modulegenes_chr$C..remanei.Gene.name_C.latens.Gene.name)

#filtering genes to get ka/ks and enc values
ortholog_enc_kaks_data_wgcna = (ortholog_enc_kaks_data[ortholog_enc_kaks_data$ortho_names %in% WGCNA_modulegenes_chr$C..remanei.Gene.name_C.latens.Gene.name, ])
WGCNA_modulegenes_filtered = (WGCNA_modulegenes_chr[WGCNA_modulegenes_chr$C..remanei.Gene.name_C.latens.Gene.name %in% ortholog_enc_kaks_data$ortho_names, ])
cumulative_enc_wgcna = (cbind(ortholog_enc_kaks_data_wgcna, WGCNA_modulegenes_filtered[, 4:7]))
names(cumulative_enc_wgcna)[11] = "Module"
# Uncomment to write the results to text files
# write.csv(cumulative_enc_wgcna, file = "../3-Output/cumulative_enc_wgcna.csv",
#             row.names = FALSE, col.names = FALSE, quote = FALSE)

```

##C. Filter dn/ds data using corrected dn/ds values

```{r}
# Calculating 99 percentile to remove outliers based on corrected dn/ds values
top_99_percentile <- quantile((cumulative_enc_wgcna %>% 
                             pull(dS_corrected)),
                          na.rm = TRUE,
                          0.99)


cumulative_enc_wgcna_filt <- cumulative_enc_wgcna %>% 
  filter(dS_corrected < top_99_percentile)

# Plotting density for filtered dn/dds_corrected data
C9_a = hist(cumulative_enc_wgcna_filt$dS_corrected, main = "Histogram of corrected ds", xlab = "ds_corrected", col = "lightblue", border = "black")

C9_b = hist(cumulative_enc_wgcna$dS_corrected, main = "Histogram of ds", xlab = "ds", col = "lightblue", border = "black", 
     xlim = c(0, 6000))



# Removing genes with very low ds values (these genes will end up having very high dn/ds ratio affecting the averages)
# 
# gene_remove = c("KAF1764021.1", "KAF1747059.1", "KAF1752828.1", "KAF1762881.1", "KAF1761359.1", "KAF1763268.1")
# 
# nrow(cumulative_enc_wgcna)
# cumulative_enc_wgcna_filt <- (cumulative_enc_wgcna[!cumulative_enc_wgcna$Remanei_Gene %in% gene_remove, ])


# # Assign Module numbers to Module names
# cumulative_enc_wgcna_filt$Module_num = "M"
# # Create a named vector for mapping
# module_map <- c("magenta" = "M1", "brown" = "M2", "greenyellow" = "M3", 
#                 "black" = "M4", "turquoise" = "M5", "green" = "M6", 
#                 "pink" = "M7", "red" = "M8", "blue" = "M9", 
#                 "yellow" = "M10", "purple" = "M11", "grey" = "M0")
# 
# # Update Module_num based on the mapping
# cumulative_enc_wgcna_filt$Module_num <- module_map[cumulative_enc_wgcna_filt$Module]


```

##D. Basic statistics to compare different modules

```{r}

# Custom mode function
get_mode <- function(v) {
  uniq_v <- unique(v)
  uniq_v[which.max(tabulate(match(v, uniq_v)))]
}

# Function to calculate average, median, and mode for each module
calculate_module_averages <- function(data, module_column, module_number, dn_ds_column, enc_column) {
  # Filter the data for the specific modules
  module_data <- data %>%
    filter(!!sym(module_column) %in% module_number)
  
  # Calculate average, median, and mode values for each module
  averages <- module_data %>%
    group_by(!!sym(module_column)) %>%
    summarise(
      average_dn_ds = mean(!!sym(dn_ds_column), na.rm = TRUE),
      average_enc = mean(!!sym(enc_column), na.rm = TRUE),
      .groups = 'drop'
    )
  
  medians <- module_data %>%
    group_by(!!sym(module_column)) %>%
    summarise(
      median_dn_ds = median(!!sym(dn_ds_column), na.rm = TRUE),
      median_enc = median(!!sym(enc_column), na.rm = TRUE),
      .groups = 'drop'
    )
  
  modes <- module_data %>%
    group_by(!!sym(module_column)) %>%
    summarise(
      mode_dn_ds = get_mode(!!sym(dn_ds_column)),
      mode_enc = get_mode(!!sym(enc_column)),
      .groups = 'drop'
    )
  
  # Return results as a list
  return(list(averages = averages, medians = medians, modes = modes))
}

# Assuming cumulative_enc_wgcna is your data frame and module_names is a vector of module names
module_number = module_number$module_number
result <- calculate_module_averages(cumulative_enc_wgcna_filt, "Module_num", module_number, "dn.ds_corrected", "ENC")

# To view results
print(result$averages)
print(result$medians)
print(result$modes)
module_stast = as.data.frame(result)
#module_stast$module_num = c("M4", "M9", "M2", "M6", "M3", "M0", "M1", "M7", "M11", "M8", "M5", "M10")

# Uncomment to write the results to text files
# write.csv(module_stast, file = "../3-Output/WGCNA/module_stats.csv",
#             row.names = FALSE, col.names = FALSE, quote = FALSE)
```

##E. Visualize data across all modules
###I. Violin Plot for dn/ds by modules
```{r, warning=FALSE}
# Calculate mean values for each module
mean_values <- cumulative_enc_wgcna_filt %>%
  group_by(Module_num) %>%
  summarize(mean_dn_ds = mean(dn.ds_corrected, na.rm = TRUE))

median_values = cumulative_enc_wgcna_filt %>%
  group_by(Module_num) %>%
  summarize(median_dn_ds = median(dn.ds_corrected, na.rm = TRUE))
# Specify the desired order of Module_num
module_order <- c("M10", "M4", "M6", "M3", "M11", "M2", "M7", "M5", "M1", "M9", "M8", "M0")

#Converting to log normalized values
cumulative_enc_wgcna_filt$log_dn_ds <- log(cumulative_enc_wgcna_filt$dn.ds_corrected + 1, base = 10)  # Add 1 to avoid log(0)

# Violin Plot for dn/ds by modules
D9_I_a = ggplot(cumulative_enc_wgcna_filt, 
       aes(x = factor(Module_num, levels = module_order), y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "Violin plot of dn/ds by Modules",
       x = "Modules",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  geom_text(data = mean_values, aes(x = Module_num, y = mean_dn_ds, 
                                     label = round(mean_dn_ds, 2)),
            vjust = -5, size = 5, color = "black") +
  # scale_x_discrete(labels = c("black" = "M4","blue" = "M9", "brown" = "M2", "green" = "M6", "greenyellow" = "M3", "grey" = "M0", "magenta" = "M1", "pink" = "M7", "purple" = "M11", "red" = "M8", "turquoise" = "M5", "yellow" = "M10"))+
  ylim(0, 1.5)+
  theme_minimal()+
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )

#Log normalized data
D9_I_b = ggplot(cumulative_enc_wgcna_filt, 
       aes(x = factor(Module_num, levels = module_order), y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "Violin plot of log dn/ds by Modules",
       x = "Modules",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  geom_text(data = mean_values, aes(x = Module_num, y = mean_dn_ds, 
                                    label = round(mean_dn_ds, 2)),
            vjust = -5, size = 5, color = "black") +
  scale_y_log10(
    labels = label_log(digits = 2), 
    breaks = c(1e-10, 1e-8, 1e-6, 1e-4, 1e-2, 1, 1e2),
    limits = c(0.0001, 100)
  ) +
  annotation_logticks(sides = "l") +
  theme_minimal()+
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
  )

#LINEAR SCALE BOXPLOT
D9_I_c = ggplot(cumulative_enc_wgcna_filt, 
       aes(x = factor(Module_num, levels = module_order), y = dn.ds_corrected)) +
  geom_boxplot(fill = "lightblue", color = "black") +
  labs(title = "Violin plot of log dn/ds by Modules",
       x = "Modules",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  geom_text(data = mean_values, aes(x = Module_num, y = mean_dn_ds, 
                                    label = round(mean_dn_ds, 2)),
            vjust = -5, size = 5, color = "black") +
  theme_minimal()+
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
  )

#LOG SCALE BOXPLOT
D9_I_d = ggplot(cumulative_enc_wgcna_filt, 
       aes(x = factor(Module_num, levels = module_order), y = dn.ds_corrected)) +
  geom_boxplot(fill = "lightblue", color = "black") +
  labs(title = "Violin plot of log dn/ds by Modules",
       x = "Modules",
       y = "dn/ds") +
  stat_summary(fun = median, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  geom_text(data = median_values, aes(x = Module_num, y = median_dn_ds, 
                                    label = round(median_dn_ds, 2)),
            vjust = -5, size = 5, color = "black") +
  theme_minimal()+
  scale_y_log10(
    labels = label_log(digits = 4), 
    breaks = c(1e-3, 1e-2, 1e-1, 1, 1e1),
    limits = c(1e-3, 1e1)
  ) +
  annotation_logticks(sides = "l") +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
  )


 
##Ascending according to median values

# Step 2: Create an ordered factor based on median
module_order <- median_values$Module_num
cumulative_enc_wgcna_filt$Module_num <- factor(cumulative_enc_wgcna_filt$Module_num, levels = module_order)
median_values$Module_num <- factor(median_values$Module_num, levels = module_order)

# Step 3: Plot
ggplot(cumulative_enc_wgcna_filt, 
       aes(x = Module_num, y = dn.ds_corrected)) +
  geom_boxplot(fill = "lightblue", color = "black") +
  labs(title = "Violin plot of log dn/ds by Modules",
       x = "Modules",
       y = "dn/ds") +
  stat_summary(fun = median, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  geom_text(data = median_values, aes(x = Module_num, y = median_dn_ds, 
                                      label = round(median_dn_ds, 2)),
            vjust = -5, size = 5, color = "black") +
  theme_minimal() +
  scale_y_log10(
    labels = scales::label_log(digits = 4), 
    breaks = c(1e-3, 1e-2, 1e-1, 1, 1e1),
    limits = c(1e-3, 1e1)
  ) +
  annotation_logticks(sides = "l") +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
  )




```

###II. Violin plot for ENC by modules
```{r}
# Calculate mean values for each module
mean_enc <- cumulative_enc_wgcna_filt %>%
  group_by(Module_num) %>%
  summarize(mean_enc = mean(ENC, na.rm = TRUE))


D9_II_a = ggplot(cumulative_enc_wgcna_filt, 
                 aes(x = factor(Module_num, levels = module_order), y = ENC)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "Violin plot of ENC by Modules",
       x = "Modules",
       y = "ENC") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  geom_text(data = mean_enc, aes(x = Module_num, y = mean_enc, 
                                     label = round(mean_enc, 2)),
            vjust = -0.5, size = 5, color = "black") +
  # scale_x_discrete(labels = c("black" = "M4","blue" = "M9", "brown" = "M2", "green" = "M6", "greenyellow" = "M3", "grey" = "M0", "magenta" = "M1", "pink" = "M7", "purple" = "M11", "red" = "M8", "turquoise" = "M5", "yellow" = "M10"))+
  #ylim(0, 5)+
  theme_minimal()+
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )

```

###III. ENC vs dn/ds
```{r, warning=FALSE}

D9_III_a = ggplot(cumulative_enc_wgcna_filt, aes(y = cumulative_enc_wgcna_filt$dn.ds_corrected, 
                                 x = cumulative_enc_wgcna_filt$ENC, 
                                 colour = factor(Module_num, levels = module_order))) +
  geom_point( size = 3) +  # Set point color and size
  labs(title = "Scatter Plot of ENC vs dn/ds",
       x = "ENC",
       y = "dn/ds") +
  geom_smooth(method = "lm", se = FALSE) + #Add linear regression
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )

```

###IV. Median ENC vs dn/ds
```{r, warning=FALSE}
#Median ENC vs dn/ds
D9_IV_a = ggplot(module_stast, aes(y = module_stast$medians.median_dn_ds, 
                                 x = module_stast$medians.median_enc, 
                                 colour = factor(module_stast$medians.Module_num, levels = module_order))) +
  geom_point( size = 3) +  # Set point color and size
  labs(title = "Scatter Plot of median ENC vs dn/ds",
       x = "ENC",
       y = "dn/ds") +
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )

#Converting to log normalized values
module_stast$log_dn_ds <- log(module_stast$averages.average_dn_ds + 1, base = 10)  # Add 1 to avoid log(0)

#Average ENC vs dn/ds
D9_IV_b = ggplot(module_stast, aes(y = module_stast$averages.average_dn_ds, 
                                 x = module_stast$averages.average_enc, 
                                 colour = factor(module_stast$averages.Module_num, levels = module_order))) +
  geom_point( size = 3) +  # Set point color and size
  labs(title = "Scatter Plot of Avg. ENC vs dn/ds",
       x = "ENC",
       y = "dn/ds") +
  theme_minimal()+
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )

D9_IV_c = ggplot(module_stast, aes(y = module_stast$log_dn_ds, 
                                 x = module_stast$averages.average_enc, 
                                 colour = factor(module_stast$averages.Module_num, levels = module_order))) +
  geom_point( size = 3) +  # Set point color and size
  labs(title = "Scatter Plot of Avg. ENC vs log dn/ds",
       x = "ENC",
       y = "log 10 dn/ds") +
  theme_minimal()+
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )

```

###V. Calculating significant differences between average dn/ds to compare modules
ANOVA and Tukey's HSD post-hoc test
```{r}
# Perform ANOVA
anova_result <- aov(dn.ds_corrected ~ Module_num, data = cumulative_enc_wgcna_filt)
summary(anova_result)

# Tukey's HSD post-hoc test
tukey_result <- TukeyHSD(anova_result)
View((tukey_result[["Module_num"]]))
tukey_Reslt_df = as.data.frame((tukey_result[["Module_num"]])) #20 significant comparisons p < 0.001

# Uncomment to write the results to text files
write.csv(tukey_Reslt_df, file = "../3-Output/WGCNA/tukey_module_comparison.csv",
            row.names = TRUE, col.names = FALSE, quote = FALSE)

# CLD letters to make tukey's values easy to display
#library(multcompView)
cld_data <- multcompLetters(tukey_result$Module_num[,"p adj"])

# Turn into data frame
cld_data_df <- data.frame(
  Module_num = names(cld_data$Letters),
  Letters = cld_data$Letters
)

# Compute y positions for label placement
y_positions_data <- cumulative_enc_wgcna_filt %>%
  group_by(Module_num) %>%
  summarise(y_pos = max(log_dn_ds) * 1.1)

cld_data_df <- left_join(cld_data_df, y_positions_data, by = "Module_num")


##Plotting with p-value highlighted
#library(ggsignif)

ggplot(cumulative_enc_wgcna_filt, aes(x = Module_num, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "Violin plot of dN/dS by Modules",
       x = "Modules",
       y = "dN/dS") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  # scale_x_discrete(labels = c("black" = "M4", "blue" = "M9", "brown" = "M2", 
  #                             "green" = "M6", "greenyellow" = "M3", "grey" = "M0", 
  #                             "magenta" = "M1", "pink" = "M7", "purple" = "M11", 
  #                             "red" = "M8", "turquoise" = "M5", "yellow" = "M10")) +
  geom_signif(comparisons = list(c("M8", "M4"), c("M4", "M10"))) +  # Replace with your comparisons
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )

```

Kruskal Wallis and Dunn's test (post hoc)
```{r}
kruskal.test(cumulative_enc_wgcna_filt$dn.ds_corrected ~ cumulative_enc_wgcna_filt$Module_num, data = cumulative_enc_wgcna_filt)

library(FSA)  # for Dunn's test
duun_test_wgcna = (dunnTest(dn.ds_corrected ~ Module_num, data = cumulative_enc_wgcna_filt, method = "bonferroni"))
dunn_test_wgcna = as.data.frame(duun_test_wgcna$res)

# Uncomment to write the results to text files
# write.csv(dunn_test_wgcna, file = "../3-Output/WGCNA/kruska_dunn_module_comparison.csv",
#             row.names = TRUE, col.names = FALSE, quote = FALSE)

```

Glm and posthoc test
```{r}
# Fit a GLM model (normal distribution with identity link function, which is default)
cumulative_enc_wgcna_filt
cumulative_enc_wgcna_filt$dnds_e <- (cumulative_enc_wgcna_filt$dn.ds_corrected + 1e-5)
wgcna_glm_dnds <- glm(cumulative_enc_wgcna_filt$dnds_e ~ cumulative_enc_wgcna_filt$Module_num, data = cumulative_enc_wgcna_filt, family = Gamma(link = "log"))

# Use the emmeans package to perform pairwise comparisons

wgcna_emm_dnds <- emmeans(wgcna_glm_dnds, pairwise ~ Module_num)
summary(wgcna_emm_dnds)
```

###VI. Organizing Plots
```{r, warning=FALSE}
# Arrange plots in a single panel
combined_dn_ds <- D9_I_d + D9_II_a + D9_III_a + D9_IV_b +plot_layout(ncol = 2)


# Saving as pdf
ggsave("../3-Output/WGCNA/cumulative_dnds_enc.pdf", plot = combined_dn_ds, width = 34, height = 15)

#Saving as png
ggsave("../3-Output/WGCNA/cumulative_dnds_enc.png", plot = combined_dn_ds, width = 34, height = 15)

```


##F. Visualize dn/ds and ENC for hub and non-hub genes
###I. Categorizing Species-bised modules into hub and non-hub and dn/ds data
####i. Yellow module (M10)
```{r}
head(yelow_gs_meme)

yello_hub = (yelow_gs_meme %>% filter(category == "hub genes")) %>% rownames_to_column()
yello_nonhub = (yelow_gs_meme %>% filter(category == "non-hub")) %>% rownames_to_column()

cumulative_enc_wgcna_yellow = cumulative_enc_wgcna_filt %>% 
  filter(Module == "yellow") %>% 
  mutate(Category = case_when(
    ortho_names %in% yello_hub$rowname ~ "Hub-genes",
    TRUE ~ "Non-hub"
  ))

## Plotting based on hub vs non-hub genes

F9_I_a = ggplot(cumulative_enc_wgcna_yellow, aes(x = cumulative_enc_wgcna_yellow$Category, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "M10 (Species-biased)",
       x = "Gene Category",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  ylim(0, 2)+
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )



mean_values_yellow <- cumulative_enc_wgcna_yellow %>%
  group_by(Category) %>%
  summarize(mean_dN_dS = mean(dn.ds_corrected, na.rm = TRUE))

# Create the violin plot with mean annotations
F9_I_b = ggplot(cumulative_enc_wgcna_yellow, aes(x = Category, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "M10 (Species-biased)",
       x = "Gene Category",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  ylim(0, 2) +
  theme_minimal() +
  geom_text(data = mean_values_yellow, aes(x = Category, y = mean_dN_dS, label = round(mean_dN_dS, 2)), 
            vjust = -0.5, color = "red")+  # Adjust vjust to position the text
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )


```

Glm and posthoc test
```{r}
# Fit a GLM model (normal distribution with identity link function, which is default)
cumulative_enc_wgcna_yellow
cumulative_enc_wgcna_yellow$dnds_e <- (cumulative_enc_wgcna_yellow$dn.ds_corrected + 1e-5)
yellow_glm_dnds <- glm(cumulative_enc_wgcna_yellow$dnds_e ~ cumulative_enc_wgcna_yellow$Category, data = cumulative_enc_wgcna_yellow, family = Gamma(link = "log"))

# Use the emmeans package to perform pairwise comparisons

yellow_emm_dnds <- emmeans(yellow_glm_dnds, pairwise ~ Category)
summary(yellow_emm_dnds)
```

####ii. Black module (M4)
```{r}

head(black_gs_meme)

black_hub = (black_gs_meme %>% filter(category == "hub genes")) %>% rownames_to_column()
black_nonhub = (black_gs_meme %>% filter(category == "non-hub")) %>% rownames_to_column()

cumulative_enc_wgcna_black = cumulative_enc_wgcna_filt %>% 
  filter(Module == "black") %>% 
  mutate(Category = case_when(
    ortho_names %in% black_hub$rowname ~ "Hub-genes",
    TRUE ~ "Non-hub"
  ))

## Plotting based on hub vs non-hub genes

F9_I_c = ggplot(cumulative_enc_wgcna_black, aes(x = cumulative_enc_wgcna_black$Category, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "M4 (Species-biased)",
       x = "Gene Category",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  ylim(0, 2)+
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )



mean_values_black <- cumulative_enc_wgcna_black %>%
  group_by(Category) %>%
  summarize(mean_dN_dS = mean(dn.ds_corrected, na.rm = TRUE))

# Create the violin plot with mean annotations
F9_I_d = ggplot(cumulative_enc_wgcna_black, aes(x = Category, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "M4 (Species-biased)",
       x = "Gene Category",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  ylim(0, 2) +
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    ) +
  geom_text(data = mean_values_black, aes(x = Category, y = mean_dN_dS, label = round(mean_dN_dS, 2)), 
            vjust = -0.5, color = "red")  # Adjust vjust to position the text

```

Glm and posthoc test
```{r}
# Fit a GLM model (normal distribution with identity link function, which is default)
cumulative_enc_wgcna_black
cumulative_enc_wgcna_black$dnds_e <- (cumulative_enc_wgcna_black$dn.ds_corrected + 1e-5)
black_glm_dnds <- glm(cumulative_enc_wgcna_black$dnds_e ~ cumulative_enc_wgcna_black$Category, data = cumulative_enc_wgcna_black, family = Gamma(link = "log"))

# Use the emmeans package to perform pairwise comparisons

black_emm_dnds <- emmeans(black_glm_dnds, pairwise ~ Category)
summary(black_emm_dnds)
```

###II. Categorizing sex-bised modules into hub and non-hub and dn/ds data
####i. Brown module (M2)
```{r}
head(brown_gs_meme)

brown_hub = (brown_gs_meme %>% filter(category == "hub genes")) %>% rownames_to_column()
brown_nonhub = (brown_gs_meme %>% filter(category == "non-hub")) %>% rownames_to_column()

cumulative_enc_wgcna_brown = cumulative_enc_wgcna_filt %>% 
  filter(Module == "brown") %>% 
  mutate(Category = case_when(
    ortho_names %in% brown_hub$rowname ~ "Hub-genes",
    TRUE ~ "Non-hub"
  ))

## Plotting based on hub vs non-hub genes

F9_II_a = ggplot(cumulative_enc_wgcna_brown, aes(x = cumulative_enc_wgcna_brown$Category, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "M2 (Sex-biased)",
       x = "Gene Category",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  ylim(0, 2)+
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )



mean_values_brown <- cumulative_enc_wgcna_brown %>%
  group_by(Category) %>%
  summarize(mean_dN_dS = mean(dn.ds_corrected, na.rm = TRUE))

# Create the violin plot with mean annotations
F9_II_b = ggplot(cumulative_enc_wgcna_brown, aes(x = Category, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "M2 (Sex-biased)",
       x = "Gene Category",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  ylim(0, 2) +
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    ) +
  geom_text(data = mean_values_brown, aes(x = Category, y = mean_dN_dS, label = round(mean_dN_dS, 2)), 
            vjust = -0.5, color = "red")  # Adjust vjust to position the text

```

Glm and posthoc test
```{r}
# Fit a GLM model (normal distribution with identity link function, which is default)
cumulative_enc_wgcna_brown
cumulative_enc_wgcna_brown$dnds_e <- (cumulative_enc_wgcna_brown$dn.ds_corrected + 1e-5)
brown_glm_dnds <- glm(cumulative_enc_wgcna_brown$dnds_e ~ cumulative_enc_wgcna_brown$Category, data = cumulative_enc_wgcna_brown, family = Gamma(link = "log"))

# Use the emmeans package to perform pairwise comparisons

brown_emm_dnds <- emmeans(brown_glm_dnds, pairwise ~ Category)
summary(brown_emm_dnds)
```

####ii. Pink module (M7)
```{r}

head(pink_gs_meme)

pink_hub = (pink_gs_meme %>% filter(category == "hub genes")) %>% rownames_to_column()
pink_nonhub = (pink_gs_meme %>% filter(category == "non-hub")) %>% rownames_to_column()

cumulative_enc_wgcna_pink = cumulative_enc_wgcna_filt %>% 
  filter(Module == "pink") %>% 
  mutate(Category = case_when(
    ortho_names %in% pink_hub$rowname ~ "Hub-genes",
    TRUE ~ "Non-hub"
  ))

## Plotting based on hub vs non-hub genes

F9_II_c = ggplot(cumulative_enc_wgcna_pink, aes(x = cumulative_enc_wgcna_pink$Category, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "M7 (Sex-biased)",
       x = "Gene Category",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  ylim(0, 2)+
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )



mean_values_pink <- cumulative_enc_wgcna_pink %>%
  group_by(Category) %>%
  summarize(mean_dN_dS = mean(dn.ds_corrected, na.rm = TRUE))

# Create the violin plot with mean annotations
F9_II_d = ggplot(cumulative_enc_wgcna_pink, aes(x = Category, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "M7 (Sex-biased)",
       x = "Gene Category",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  ylim(0, 2) +
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    ) +
  geom_text(data = mean_values_pink, aes(x = Category, y = mean_dN_dS, label = round(mean_dN_dS, 2)), 
            vjust = -0.5, color = "red")  # Adjust vjust to position the text

```

Glm and posthoc test
```{r}
# Fit a GLM model (normal distribution with identity link function, which is default)
cumulative_enc_wgcna_pink
cumulative_enc_wgcna_pink$dnds_e <- (cumulative_enc_wgcna_pink$dn.ds_corrected + 1e-5)
pink_glm_dnds <- glm(cumulative_enc_wgcna_pink$dnds_e ~ cumulative_enc_wgcna_pink$Category, data = cumulative_enc_wgcna_pink, family = Gamma(link = "log"))

# Use the emmeans package to perform pairwise comparisons

pink_emm_dnds <- emmeans(pink_glm_dnds, pairwise ~ Category)
summary(pink_emm_dnds)
```

###III. Categorizing TISSUE-bised modules into hub and non-hub and dn/ds data
####i. Magenta module (M1)
```{r}
head(magenta_gs_meme)

magenta_hub = (magenta_gs_meme %>% filter(category == "hub genes")) %>% rownames_to_column()
magenta_nonhub = (magenta_gs_meme %>% filter(category == "non-hub")) %>% rownames_to_column()

cumulative_enc_wgcna_magenta = cumulative_enc_wgcna_filt %>% 
  filter(Module == "magenta") %>% 
  mutate(Category = case_when(
    ortho_names %in% magenta_hub$rowname ~ "Hub-genes",
    TRUE ~ "Non-hub"
  ))

## Plotting based on hub vs non-hub genes

F9_III_a = ggplot(cumulative_enc_wgcna_magenta, aes(x = cumulative_enc_wgcna_magenta$Category, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "M1 (Tissue-biased)",
       x = "Gene Category",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  ylim(0, 2)+
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )



mean_values_magenta <- cumulative_enc_wgcna_magenta %>%
  group_by(Category) %>%
  summarize(mean_dN_dS = mean(dn.ds_corrected, na.rm = TRUE))

# Create the violin plot with mean annotations
F9_III_b = ggplot(cumulative_enc_wgcna_magenta, aes(x = Category, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "M1 (Tissue-biased)",
       x = "Gene Category",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  ylim(0, 2) +
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    ) + 
  geom_text(data = mean_values_magenta, aes(x = Category, y = mean_dN_dS, label = round(mean_dN_dS, 2)), 
            vjust = -0.5, color = "red")  # Adjust vjust to position the text

```

Glm and posthoc test
```{r}
# Fit a GLM model (normal distribution with identity link function, which is default)
cumulative_enc_wgcna_magenta
cumulative_enc_wgcna_magenta$dnds_e <- (cumulative_enc_wgcna_magenta$dn.ds_corrected + 1e-5)
magenta_glm_dnds <- glm(cumulative_enc_wgcna_magenta$dnds_e ~ cumulative_enc_wgcna_magenta$Category, data = cumulative_enc_wgcna_magenta, family = Gamma(link = "log"))

# Use the emmeans package to perform pairwise comparisons

magenta_emm_dnds <- emmeans(magenta_glm_dnds, pairwise ~ Category)
summary(magenta_emm_dnds)
```

####ii. Blue module (M9)
```{r}

head(blue_gs_meme)

blue_hub = (blue_gs_meme %>% filter(category == "hub genes")) %>% rownames_to_column()
blue_nonhub = (blue_gs_meme %>% filter(category == "non-hub")) %>% rownames_to_column()

cumulative_enc_wgcna_blue = cumulative_enc_wgcna_filt %>% 
  filter(Module == "blue") %>% 
  mutate(Category = case_when(
    ortho_names %in% blue_hub$rowname ~ "Hub-genes",
    TRUE ~ "Non-hub"
  ))

## Plotting based on hub vs non-hub genes

F9_III_c = ggplot(cumulative_enc_wgcna_blue, aes(x = cumulative_enc_wgcna_blue$Category, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "M9 (Tissue-biased)",
       x = "Gene Category",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  ylim(0, 2)+
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )



mean_values_blue <- cumulative_enc_wgcna_blue %>%
  group_by(Category) %>%
  summarize(mean_dN_dS = mean(dn.ds_corrected, na.rm = TRUE))

# Create the violin plot with mean annotations
F9_III_d = ggplot(cumulative_enc_wgcna_blue, aes(x = Category, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "M9 (Tissue-biased)",
       x = "Gene Category",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  ylim(0, 2) +
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    ) + 
  geom_text(data = mean_values_blue, aes(x = Category, y = mean_dN_dS, label = round(mean_dN_dS, 2)), 
            vjust = -0.5, color = "red")  # Adjust vjust to position the text

```

Glm and posthoc test
```{r}
# Fit a GLM model (normal distribution with identity link function, which is default)
cumulative_enc_wgcna_blue
cumulative_enc_wgcna_blue$dnds_e <- (cumulative_enc_wgcna_blue$dn.ds_corrected + 1e-5)
blue_glm_dnds <- glm(cumulative_enc_wgcna_blue$dnds_e ~ cumulative_enc_wgcna_blue$Category, data = cumulative_enc_wgcna_blue, family = Gamma(link = "log"))

# Use the emmeans package to perform pairwise comparisons

blue_emm_dnds <- emmeans(blue_glm_dnds, pairwise ~ Category)
summary(blue_emm_dnds)
```

###IV. Organizing Plots
```{r, warning=FALSE}
# Arrange plots in a single panel
dn_ds_hub_nonhub <- F9_I_b + F9_I_d + F9_II_b + F9_II_d + F9_III_b + F9_III_d +plot_layout(ncol = 2)


# Saving as pdf
# ggsave("../3-Output/WGCNA/dn_ds_hub_nonhub.pdf", plot = dn_ds_hub_nonhub, width = 34, height = 15)
# 
# #Saving as png
# ggsave("../3-Output/WGCNA/dn_ds_hub_nonhub.png", plot = dn_ds_hub_nonhub, width = 34, height = 15)

```
