---
title: "6. Figures_WGCNA_Cre"
author: "Athmaja Viswanath"
date: "2024-11-12"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#0. Loading required packages
```{r}
library(WGCNA)
library(DESeq2)
library(tidyverse)
library(ggplot2)
library(ggstats)
library(CorLevelPlot)
library(gridExtra)
library(VennDiagram)
library(grid)
library(lme4)
library(dplyr)
library(ggsignif)
library(patchwork)

#`%notin%` <- Negate(`%in%`)
options(scipen = 999)  # Increase penalty to avoid scientific notation

```

```{r}
# module name information
module_number_cre <- data.frame(Module_Colour = c("MEgrey", "MEmagenta", "MEbrown", "MEblack", "MEturquoise",
                                                  "MEgreen", "MEpink", "MEred", "MEblue", "MEyellow"),
                                mcolour = c("grey", "magenta", "brown", "black", "turquoise",
                                                  "green", "pink", "red", "blue", "yellow"),
                                module_number = c("CR0", "CR4", "CR9", "CR2", "CR5", 
                                                  "CR1", "CR6", "CR7", "CR3", "CR8"))

```

#1. Plotting dendrogram 
this needs the bwnet_cre (network constructed in 5.WGCNA_Cre.Rmd), this code is also present after network creation

```{r}
# Plot the dendrogram with color assignments for both unmerged and merged modules
# plotDendroAndColors(bwnet_cre$dendrograms[[1]],
#                     cbind(bwnet_cre$unmergedColors, bwnet_cre$colors),
#                     c("unmerged", "merged"),
#                     dendroLabels = FALSE,
#                     oddGuide = TRUE,
#                     hang = 0.03,
#                     guideHang = 0.05)
# 
# 
# dendo1 = plotDendroAndColors(bwnet_cre$dendrograms[[1]],
#                              cbind(bwnet_Cre$colors),
#                              c("merged"),
#                              dendroLabels = FALSE,
#                              oddGuide = TRUE,
#                              hang = 0.03,
#                              guideHang = 0.05)
```

#2. Eigengene dendrogram

```{r}

module_eigengenes_cre = read.csv("../3-Output/WGCNA/module eigengenes values_Cre.csv")
row.names(module_eigengenes_cre) = module_eigengenes_cre$X
module_eigengenes_cre = module_eigengenes_cre[, -1]

#Eigengene Dendrograms 
eingen_dendo_cre = plotEigengeneNetworks(module_eigengenes_cre, "Eigengene dendrogram", marDendro = c(0,4,2,0),
                      plotHeatmaps = FALSE, excludeGrey = FALSE)

#Adjacency Matrix
eingen_heatmap_cre = plotEigengeneNetworks(module_eigengenes_cre, "Eigengene adjacency heatmap", marHeatmap = c(3,4,2,2),
                      plotDendrograms = FALSE, xLabelsAngle = 90, colorLabels = FALSE)

```

#3. Visualize the number of genes in each module

```{r}
###loading file which contains genes and their corresponding modules (obtained from above)
bwnet_df_cre = read.table("../3-Output/WGCNA/WGCNA_modulegenes_Cre.txt", fill = T)

# Count the number of genes in each module
modules_counts_cre = as.data.frame(table(bwnet_df_cre$V2))

# Assuming modules_counts is your data frame
modules_counts_cre$Module_num <- module_number$module_number[match(modules_counts$Var1, module_number$mcolour)]


# Create a bar plot of gene counts per module
ggplot(modules_counts, aes(x=Module_num, y = Freq, fill = Module_num))+
  geom_bar(stat = "identity") +
  xlab(" ")+ 
  ylab("Number of genes")+
  theme(plot.title = element_text(size = 26, face = "bold", hjust = 0.5, vjust = 0.5))+
  coord_cartesian(ylim = c(0, 3500))+
  #scale_fill_manual(values = c( "#afced0", "#4b7d81", "#696967", '#DCDDDF'))+
  theme(axis.title.y = element_text(size = 14, hjust = 0.5))+
  theme(axis.text.x = element_text(size = 10,face = "bold", colour = "black"))+
  theme(axis.text.y = element_text(size = 10))+
  theme(axis.title.x = element_text(size = 14))+
  theme(axis.title.y = element_text(size = 20, hjust = 0.5))+
  theme(axis.text.y = element_text(color="black", size=15))+
  guides(fill=guide_legend(title=" "))+
  # scale_colour_brewer(palette = 2)+
  #scale_x_discrete(limits = c("Conserved","Differentially expressed", "C. nigoni dominant", "Ambiguous"))+ ##reordering character x-axis
  expand_limits(y=0)+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)), breaks=seq(0,3500,500))+
  ggtitle("WGCNA Modules")


# Save the plot
#ggsave("../3-Output/2.1WGCNA_Modules_Plot.pdf", plot, width = 10, height = 6, dpi = 300)
```

#4. Visualize chromosomal distribution of modules
##A. Getting and plotting wgcna data across chromosomes
```{r}
#Load file with orthologous gene names and their corresponding chromosomes
cre_chr= read.table(file = "../1-Input/Cremanie_genenames_and_chromosome.txt", header = FALSE)


#Read module genes information 
WGCNA_modulegenes_cre = read.table(file = "../3-Output/WGCNA/WGCNA_modulegenes_Cre.txt")

#changing rownames to gene names (first column and deleting the first column)
rownames(WGCNA_modulegenes_cre) = WGCNA_modulegenes_cre$V1 
#WGCNA_modulegenes_cre = WGCNA_modulegenes_cre[, -1]
head(WGCNA_modulegenes_cre)


# Create a dataframe with non-NA entries based on orthologs_chr
WGCNA_modulegenes_cre_a <- WGCNA_modulegenes_cre[cre_chr$V2, ] %>%
  na.omit() 
  #rownames_to_column()
head(WGCNA_modulegenes_cre_a)

# Update rownames in cre_chr
rownames(cre_chr) <- cre_chr$V2

# Combine dataframes for genes present in both datasets - modules and chromosomal info
WGCNA_modulegenes_chr_cre <- na.omit(cre_chr[WGCNA_modulegenes_cre_a$V1, ]) %>%
  cbind(WGCNA_modulegenes_cre_a) %>%  # Combine the data
  { 
    colnames(.) <- make.unique(colnames(.))  # Ensure column names are unique
    .  # Return the modified dataframe
  } %>%
  select(-V2)  # Remove V1 column

# View the resulting data
colnames(WGCNA_modulegenes_chr_cre) = c("chromosome", "gene","module")
View(WGCNA_modulegenes_chr_cre)

WGCNA_modulegenes_chr_cre <- WGCNA_modulegenes_chr_cre %>% 
  filter(chromosome == "I" | chromosome == "II" | chromosome == "III" | chromosome == "IV" | chromosome == "V" | chromosome == "X")

#Adding Autosome vs X column
WGCNA_modulegenes_chr_cre = WGCNA_modulegenes_chr_cre %>%
  mutate(A_X = ifelse(chromosome %in% c("I", "II", "III", "IV", "V"), "Autosome", "X"))

# adding module numbers 
# Add the Module_num column to WGCNA_modulegenes_chr_cre by matching 'module' with 'mcolour'
WGCNA_modulegenes_chr_cre$Module_num <- module_number_cre$module_number[match(WGCNA_modulegenes_chr_cre$module, module_number_cre$mcolour)]

# View the updated dataframe
View(WGCNA_modulegenes_chr_cre)


WGCNA_modulegenes_chr_freq_cre <- as.data.frame(table(WGCNA_modulegenes_chr_cre$chromosome, WGCNA_modulegenes_chr_cre$Module_num))
WGCNA_modulegenes_A_X_freq_cre <- as.data.frame(table(WGCNA_modulegenes_chr_cre$A_X, WGCNA_modulegenes_chr_cre$Module_num))

View(WGCNA_modulegenes_chr_freq_cre)
View(WGCNA_modulegenes_A_X_freq_cre)

# Assigning module numbers
WGCNA_modulegenes_chr_freq$Module_num <- module_number$module_number[match(WGCNA_modulegenes_chr_freq$Var2, module_number$mcolour)]

WGCNA_modulegenes_A_X_freq$Module_num <- module_number$module_number[match(WGCNA_modulegenes_A_X_freq$Var2, module_number$mcolour)]


# Specify the desired order of Module_num
module_order_cre <- c("CR0", "CR1", "CR2", "CR3", "CR4", "CR5", "CR6", "CR7", "CR8", "CR9")

##Plotting bar plot across all chromosomes

A_4 = ggplot(WGCNA_modulegenes_chr_cre, aes(x = WGCNA_modulegenes_chr_cre$chromosome, fill = factor(Module_num, levels = module_order_cre))) + 
  geom_bar() +
  labs(x = "", y = "Number of genes", title = "Distribution of WGCNA modules") +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 20, hjust = 0.5),
    axis.text.x = element_text(size = 26, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 15, color = "black"),
    axis.title.x = element_text(size = 14)) +
  expand_limits(y = 0) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)), breaks = seq(0, 1000, 100)) +
  guides(fill = guide_legend(title = " ")) +
  labs(title = "Module Genes across Chromosomes", y = "Number of genes", x = "Chromosome") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


table(WGCNA_modulegenes_chr_cre$chromosome, WGCNA_modulegenes_chr_cre$Module_num)

# Proportion of modules on all chromosomes

B_4 = ggplot(WGCNA_modulegenes_chr_freq_cre) +
  aes(x = Var1, fill = factor(Var2, levels = module_order_cre), weight = Freq, by = Var1) +
  geom_bar(position = "fill") +
  geom_text(stat = "prop", position = position_fill(.5))+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)), breaks=seq(0,1,0.25)) +
  labs(title = "Module Proportion across Chromosomes", y = "Number of genes", x = "Chromosome") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


##Plotting bar plot across Autosomes vs X-chromsome

C_4 = ggplot(WGCNA_modulegenes_chr_cre, aes(x = WGCNA_modulegenes_chr_cre$A_X, fill = factor(Module_num, levels = module_order_cre))) + 
  geom_bar() +
  labs(x = "", y = "Number of genes", title = "Distribution of WGCNA modules") +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 20, hjust = 0.5),
    axis.text.x = element_text(size = 26, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 15, color = "black"),
    axis.title.x = element_text(size = 14)) +
  expand_limits(y = 0) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)), breaks = seq(0, 10000, 500)) +
  guides(fill = guide_legend(title = " ")) +
  labs(title = "Module Proportion across autosome and X Chromosomes", y = "Number of genes", x = "Chromosome") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  guides(fill = "none")  # Remove legend

table(WGCNA_modulegenes_chr_cre$chromosome, WGCNA_modulegenes_chr_cre$Module_num)

# Proportion of modules on all chromosomes

D_4 = ggplot(WGCNA_modulegenes_A_X_freq_cre) +
  aes(x = Var1, fill = factor(Var2, levels = module_order_cre), weight = Freq, by = Var1) +
  geom_bar(position = "fill") +
  geom_text(stat = "prop", position = position_fill(.5))+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)), breaks=seq(0,1,0.25)) +
  labs(title = "Module Proportion across autosome and X Chromosomes", y = "Proportion", x = "Chromosome") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  guides(fill = "none")  # Remove legend


```

##B Performing Enrichment analysis (hypergeometric test)

```{r}
# Dataframe with gene names, modules, and chromosome information
gene_info <- data.frame(
  gene = c("gene1", "gene2", "gene3", "gene4", "gene5", 
           "gene6", "gene7", "gene8", "gene9"),
  module = c("Module1", "Module1", "Module1", "Module2", "Module2", 
             "Module3", "Module3", "Module3", "Module3"),
  chromosome = c("a", "a", "A", "a", "A", "A", "a", "A", "a")
)
#actual data
names(WGCNA_modulegenes_chr_cre)

# Function to check enrichment for all modules in both chromosomes
check_enrichment_all <- function(WGCNA_modulegenes_chr) {
  results <- data.frame(Module = character(), Chromosome = character(), P_Value = numeric())
  
  unique_modules <- unique(WGCNA_modulegenes_chr$Module_num)
  
  for (module_name in unique_modules) {
    for (chromosome in c("X", "Autosome")) {
      module_genes <- WGCNA_modulegenes_chr$C..remanei.Gene.name_C.latens.Gene.name[WGCNA_modulegenes_chr$Module_num == module_name]
      
      # Count total genes in the specified chromosome
      total_genes_chromosome <- sum(WGCNA_modulegenes_chr$A_X == chromosome)
      total_genes_other <- nrow(WGCNA_modulegenes_chr) - total_genes_chromosome
      
      # Number of genes in the module that are in the specified chromosome
      genes_in_module <- sum(module_genes %in% WGCNA_modulegenes_chr$C..remanei.Gene.name_C.latens.Gene.name[WGCNA_modulegenes_chr$A_X == chromosome])
      
      # Hypergeometric test
      p_value <- phyper(genes_in_module - 1, total_genes_chromosome, total_genes_other, length(module_genes), lower.tail = FALSE)
      
      results <- rbind(results, data.frame(Module = module_name, Chromosome = chromosome, P_Value = p_value))
    }
  }
  
  return(results)
}


# Check enrichment for all modules
enrichment_results_cre <- check_enrichment_all(WGCNA_modulegenes_chr_cre)

# Plot the results using ggplot2 and scales
library(scales)  
E_4 = ggplot(enrichment_results_cre, aes(x = Module, y = P_Value, fill = Chromosome)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_hline(yintercept = 0.05, linetype = "dashed", color = "red") +  # Significance line at p = 0.05
  scale_y_continuous(trans = 'log10', breaks = scales::trans_breaks("log10", function(x) 10^x), labels = scales::trans_format("log10", math_format(10^.x))) +
  labs(title = "Module Enrichment in Chromosomes", y = "P-Value (log scale)", x = "Module") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))



```

##C. Organizing chromosomal plots
```{r}
# Arrange plots in a single panel
chromosomal_modules <- A_4+ B_4 + C_4 + D_4 +E_4 + plot_layout(ncol = 2)


# Saving as pdf
ggsave("../3-Output/WGCNA/Modules across chromosomes.pdf", plot = chromosomal_modules, width = 34, height = 15)

#Saving as png
ggsave("../3-Output/WGCNA/Modules across chromosomes.png", plot = chromosomal_modules, width = 34, height = 15)

```

#5. NEED TO DO THIS Visualize Eigengene Plots for each module
```{r}
#Make a text file with average module eigen gene values across replicates for each module

# Load average counts data
average_counts_cre <- read.table("../3-Output/WGCNA/Average_MEvalues.txt", sep = "\t", header = TRUE, comment.char = "#")

# Add metadata columns
average_counts$Species <- rep(c("Cre", "Clat"), each = 4)
average_counts$Sex <- rep(c("F", "F", "M", "M"), times = 2)
average_counts$Tissue <- rep(c("G", "S"), times = 4)
average_counts$S_Sp <- paste(average_counts$Species, average_counts$Sex, sep = "_")

# Display structure of the data
str(average_counts)

######################
# Plot for each module
######################

# Turquoise
average_counts %>% 
  ggplot(aes(Tissue, MEturquoise, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  stat_smooth(method = "lm") +
  ggtitle("M5 N = 3156")


# BLUE
average_counts %>% 
  ggplot(aes(Tissue, MEblue, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  stat_smooth(method = "lm") +
  ggtitle("M9 N = 2761")

# GREENYELLOW
average_counts %>% 
  ggplot(aes(Tissue, MEgreenyellow, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M3 N = 133")

# PURPLE
average_counts %>% 
  ggplot(aes(Tissue, MEpurple, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M11 N = 164")

# RED
average_counts %>% 
  ggplot(aes(Tissue, MEred, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M8 N = 652")

# BLACK
average_counts %>% 
  ggplot(aes(Tissue, MEblack, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M4 N = 583")

# GREEN
average_counts %>%
  ggplot(aes(Tissue, MEgreen, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M6 N = 822")


# YELLOW
average_counts %>% 
  ggplot(aes(Tissue, MEyellow, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M10 N = 979")

# PINK
average_counts %>% 
  ggplot(aes(Tissue, MEpink, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M7 N = 398")

# BROWN
average_counts %>% 
  ggplot(aes(Tissue, MEbrown, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M2 N = 1770")

# MAGENTA
average_counts %>% 
  ggplot(aes(Tissue, MEmagenta, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M1 N = 274")


# GREY
average_counts %>% 
  ggplot(aes(Tissue, MEgrey, col = S_Sp, group = S_Sp)) +
  geom_hline(yintercept=0, color = "black", linetype="dashed")+
  scale_y_continuous(limits = c(-0.6, 0.6)) +
  geom_point(show.legend=T, alpha=1, size = 2.5) +
  theme_bw() +
  #facet_wrap(~Species) +
  stat_smooth(method = "lm") +
  ggtitle("M0 N = 112")

```

#6. Visualize module trait association as a heatmap

```{r}
heatmap_data_cre = merge(module_eigengenes_cre, traits_cre, by = "row.names")
heatmap_data_cre = heatmap_data_cre %>% 
  column_to_rownames(var = "Row.names")

# Create a named vector for mapping
module_map_cre <- c("MEmagenta" = "CR4", "MEbrown" = "CR9", "MEblack" = "CR2", "MEturquoise" = "CR5", 
                "MEgreen" = "CR1", "MEpink" = "CR6", "MEred" = "CR7", "MEblue" = "CR3", 
                "MEyellow" = "CR8", "MEgrey" = "CR0", 
                "data.M.vs.F" = "Males vs Females", "S.vs.G" = "Soma vs Gonad")

# Rename columns using the module_map
colnames(heatmap_data_cre) <- module_map_cre[colnames(heatmap_data_cre)]

colnames(heatmap_data_cre)
a6 = CorLevelPlot(heatmap_data_cre,
             x = names(heatmap_data_cre)[11], 
             y = names(heatmap_data_cre)[1:10],
             col = c("red", "pink","white", "pink", "red"),
             main = "SEX")

b6 = CorLevelPlot(heatmap_data_cre,
             x = names(heatmap_data_cre)[12], 
             y = names(heatmap_data_cre)[1:10],
             col = c("#004b8d","skyblue", "white", "skyblue", "#004b8d" ),
             main = "TISSUES")

d6 = CorLevelPlot(heatmap_data_cre,
             x = names(heatmap_data_cre)[11:12], 
             y = names(heatmap_data_cre)[1:10],
             col = c("blue1", "skyblue", "white", "pink", "red" ), 
             main = "All Conditions")

###Modules in red/blue are significantly associated with one of the sex/species over the other
```


#9. Trends associated with effective number of codons (ENC) and dn/ds
##A. Preparing data to filter for genes with correct gene names along with enc, dn/ka and ds/ks values

```{r}
# Read data from text files
remanei_gene_names = read.table("../1-Input/remanei_gene_names.txt", fill = T)
ka_ks_enc_data = read.table("../1-Input/FitMG94_dN_dS_ENC_corrected_dS_values_all_orthologs_remanei_PX506_latens_PX534.txt", header = T)

# Check the structure data
View(ka_ks_enc_data) #14764 genes
View(remanei_gene_names) #15276 genes

#Calculate dn/ds_corrected using ds_corrected for ENC value

ka_ks_enc_data$dn.ds_corrected = ka_ks_enc_data$dN/ka_ks_enc_data$dS_corrected
#################################################################################

# Read in the C. remanei count data
cre_allcounts <- read.table("../1-Input/C.remanei_cumulative.txt", sep="\t", header=TRUE, row.names=1, comment.char = "#")

# Add row names as a column
cre_allcounts <- rownames_to_column(cre_allcounts)

# Define "%notin%" operator to negate "%in%" 
"%notin%" <- Negate("%in%")

# Remove rows from cre_allcounts where rownames match orthologs' C. remanei Gene names
cre_allcounts_sps <- cre_allcounts[cre_allcounts$rowname %notin% orthologs$C..remanei.Gene.name, ]

cre_allcounts_sps = row.names(cre_allcounts_sps)
#######################################################
#Filter Remanei gene names to retain only species-specific genes

cre_sp_genes = remanei_gene_names %>%
         filter(V4 %in% cre_allcounts_sps$rowname) 

# Filter ENC data for genes with matching Remanei names
enc_data_cre <- ka_ks_enc_data %>%
  filter(Remanei_Gene %in% cre_sp_genes$V1) 

# Filter Remanei gene names based on the filtered ENC data
remanei_filtered_names_cre <- cre_sp_genes %>%
  filter(V1 %in% enc_data_cre$Remanei_Gene) 

# Combine the filtered ENC data with corresponding Remanei names
enc_data_cre <- cbind(enc_data_cre, remanei_filtered_names_cre)
nrow(enc_data_cre) #932 genes



# Save the data

# Uncomment to write the results to text files
# write.table(remanei_filtered_names_cre, file = "../3-Output/species_specific_crenames.txt",
#               row.names = FALSE, col.names = TRUE, quote = FALSE)
#  
# write.table(enc_data_cre, file = "../3-Output/enc_data_kaks_cre.txt",
#              row.names = FALSE, col.names = TRUE, quote = FALSE)
```

##B. Filter ENC data for diffferent modules
```{r}
#Using chromosomal and module information from section #4
str(WGCNA_modulegenes_chr_cre)

# Use the previously made ka, ks, enc data
str(enc_data_cre)

#creatig ga new columns
enc_data_cre = enc_data_cre %>%
  arrange(enc_data_cre$V4)

WGCNA_modulegenes_chr_cre = WGCNA_modulegenes_chr_cre %>%
  arrange(WGCNA_modulegenes_chr_cre$gene)

#filtering genes to get ka/ks and enc values
enc_kaks_data_wgcna_cre = (enc_data_cre[enc_data_cre$V4 %in% WGCNA_modulegenes_chr_cre$gene, ])

WGCNA_modulegenes_filtered_cre = (WGCNA_modulegenes_chr_cre[WGCNA_modulegenes_chr_cre$gene %in% enc_data_cre$V4, ])
cumulative_enc_wgcna_cre = (cbind(enc_kaks_data_wgcna_cre, WGCNA_modulegenes_filtered_cre))

# Uncomment to write the results to text files
# write.csv(cumulative_enc_wgcna_cre, file = "../3-Output/cumulative_enc_wgcna_cre.csv",
#             row.names = FALSE, col.names = FALSE, quote = FALSE)

```

##C. Filter dn/ds data using corrected dn/ds values

```{r}
# Calculating 95 percentile to remove outliers based on corrected dn/ds values
top_99_percentile_cre <- quantile((cumulative_enc_wgcna_cre %>% 
                             pull(dS_corrected)),
                          na.rm = TRUE,
                          0.99)


cumulative_enc_wgcna_filt_cre <- cumulative_enc_wgcna_cre %>% 
  filter(dS_corrected < top_99_percentile_cre)

# Plotting density for filtered dn/dds_corrected data
C9_a = hist(cumulative_enc_wgcna_cre$dS_corrected, main = "Histogram of corrected ds", xlab = "ds_corrected", col = "lightblue", border = "black")

C9_b = hist(cumulative_enc_wgcna_cre$dS, main = "Histogram of ds", xlab = "ds", col = "lightblue", border = "black", 
     xlim = c(0, 6000))



# Removing genes with very low ds values (these genes will end up having very high dn/ds ratio affecting the averages)
# 
# gene_remove = c("KAF1764021.1", "KAF1747059.1", "KAF1752828.1", "KAF1762881.1", "KAF1761359.1", "KAF1763268.1")
# 
# nrow(cumulative_enc_wgcna)
# cumulative_enc_wgcna_filt <- (cumulative_enc_wgcna[!cumulative_enc_wgcna$Remanei_Gene %in% gene_remove, ])

```

##C. Basic statistics to compare different modules

```{r}

# Custom mode function
get_mode <- function(v) {
  uniq_v <- unique(v)
  uniq_v[which.max(tabulate(match(v, uniq_v)))]
}

# Function to calculate average, median, and mode for each module
calculate_module_averages <- function(data, module_column, module_names, dn_ds_column, enc_column) {
  # Filter the data for the specific modules
  module_data <- data %>%
    filter(!!sym(module_column) %in% module_names)
  
  # Calculate average, median, and mode values for each module
  averages <- module_data %>%
    group_by(!!sym(module_column)) %>%
    summarise(
      average_dn_ds = mean(!!sym(dn_ds_column), na.rm = TRUE),
      average_enc = mean(!!sym(enc_column), na.rm = TRUE),
      .groups = 'drop'
    )
  
  medians <- module_data %>%
    group_by(!!sym(module_column)) %>%
    summarise(
      median_dn_ds = median(!!sym(dn_ds_column), na.rm = TRUE),
      median_enc = median(!!sym(enc_column), na.rm = TRUE),
      .groups = 'drop'
    )
  
  modes <- module_data %>%
    group_by(!!sym(module_column)) %>%
    summarise(
      mode_dn_ds = get_mode(!!sym(dn_ds_column)),
      mode_enc = get_mode(!!sym(enc_column)),
      .groups = 'drop'
    )
  
  # Return results as a list
  return(list(averages = averages, medians = medians, modes = modes))
}

# Assuming cumulative_enc_wgcna is your data frame and module_names is a vector of module names
module_num_cre = module_number_cre$module_number
result_cre <- calculate_module_averages(cumulative_enc_wgcna_filt_cre, "Module_num", module_num_cre, "dn.ds_corrected", "ENC")

# To view results
print(result_cre$averages)
print(result_cre$medians)
print(result_cre$modes)
module_stats_cre = as.data.frame(result_cre)

# Uncomment to write the results to text files
# write.csv(module_stats_cre, file = "../3-Output/WGCNA/module_stats_cre.csv",
#             row.names = FALSE, col.names = FALSE, quote = FALSE)
```

##D. Visualize data across different modules
###I. Violin Plot for dn/ds by modules
```{r, warning=FALSE}
# Calculate mean values for each module
mean_values_cre <- cumulative_enc_wgcna_filt_cre %>%
  group_by(Module_num) %>%
  summarize(mean_dn_ds = mean(dn.ds_corrected, na.rm = TRUE))

# Specify the desired order of Module_num
module_order_cre <- c("CR0", "CR1", "CR2", "CR3", "CR4", "CR5", "CR6", "CR7", "CR8", "CR9")

#Converting to log normalized values
cumulative_enc_wgcna_filt_cre$log_dn_ds <- log(cumulative_enc_wgcna_filt_cre$dn.ds_corrected + 1, base = 10)  # Add 1 to avoid log(0)

# Violin Plot for dn/ds by modules
D9_I_a = ggplot(cumulative_enc_wgcna_filt_cre, 
       aes(x = factor(Module_num, levels = module_order_cre), y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "Violin plot of dn/ds by Modules",
       x = "Modules",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  geom_text(data = mean_values_cre, aes(x = Module_num, y = mean_dn_ds, 
                                     label = round(mean_dn_ds, 2)),
            vjust = -5, size = 5, color = "black") +
  # scale_x_discrete(labels = c("black" = "M4","blue" = "M9", "brown" = "M2", "green" = "M6", "greenyellow" = "M3", "grey" = "M0", "magenta" = "M1", "pink" = "M7", "purple" = "M11", "red" = "M8", "turquoise" = "M5", "yellow" = "M10"))+
  ylim(0, 1.5)+
  theme_minimal()+
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )

#Log normalized data
D9_I_b = ggplot(cumulative_enc_wgcna_filt_cre, 
       aes(x = factor(Module_num, levels = module_order_cre), y = log_dn_ds)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "Violin plot of log dn/ds by Modules",
       x = "Modules",
       y = "dn/ds") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  geom_text(data = mean_values_cre, aes(x = Module_num, y = mean_dn_ds, 
                                     label = round(mean_dn_ds, 2)),
            vjust = -5, size = 5, color = "black") +
  # scale_x_discrete(labels = c("black" = "M4","blue" = "M9", "brown" = "M2", "green" = "M6", "greenyellow" = "M3", "grey" = "M0", "magenta" = "M1", "pink" = "M7", "purple" = "M11", "red" = "M8", "turquoise" = "M5", "yellow" = "M10"))+
  #ylim(0, 1)+
  theme_minimal()+
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )
```

###II. Violin plot for ENC by modules
```{r}
# Calculate mean values for each module
mean_enc_cre <- cumulative_enc_wgcna_filt_cre %>%
  group_by(Module_num) %>%
  summarize(mean_enc = mean(ENC, na.rm = TRUE))


D9_II_a = ggplot(cumulative_enc_wgcna_filt_cre, 
                 aes(x = factor(Module_num, levels = module_order_cre), y = ENC)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "Violin plot of ENC by Modules",
       x = "Modules",
       y = "ENC") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  geom_text(data = mean_enc_cre, aes(x = Module_num, y = mean_enc, 
                                     label = round(mean_enc, 2)),
            vjust = -0.5, size = 5, color = "black") +
  # scale_x_discrete(labels = c("black" = "M4","blue" = "M9", "brown" = "M2", "green" = "M6", "greenyellow" = "M3", "grey" = "M0", "magenta" = "M1", "pink" = "M7", "purple" = "M11", "red" = "M8", "turquoise" = "M5", "yellow" = "M10"))+
  #ylim(0, 5)+
  theme_minimal()+
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )



```

###III. ENC vs dn/ds
```{r, warning=FALSE}

D9_III_a = ggplot(cumulative_enc_wgcna_filt_cre, aes(y = cumulative_enc_wgcna_filt_cre$dn.ds_corrected, 
                                 x = cumulative_enc_wgcna_filt_cre$ENC, 
                                 colour = factor(Module_num, levels = module_order_cre))) +
  geom_point( size = 3) +  # Set point color and size
  labs(title = "Scatter Plot of ENC vs dn/ds",
       x = "ENC",
       y = "dn/ds") +
  geom_smooth(method = "lm", se = FALSE) + #Add linear regression
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )

```
###IV. Median ENC vs dn/ds
```{r, warning=FALSE}
#Median ENC vs dn/ds
D9_IV_a = ggplot(module_stats_cre, aes(y = module_stats_cre$medians.median_dn_ds, 
                                 x = module_stats_cre$medians.median_enc, 
                                 colour = factor(module_stats_cre$averages.Module_num, levels = module_order_cre))) +
  geom_point( size = 3) +  # Set point color and size
  labs(title = "Scatter Plot of median ENC vs dn/ds",
       x = "ENC",
       y = "dn/ds") +
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )

#Converting to log normalized values
module_stats_cre$log_dn_ds <- log(module_stats_cre$averages.average_dn_ds + 1, base = 10)  # Add 1 to avoid log(0)

#Average ENC vs dn/ds
D9_IV_b = ggplot(module_stats_cre, aes(y = module_stats_cre$averages.average_dn_ds, 
                                 x = module_stats_cre$averages.average_enc, 
                                 colour = factor(module_stats_cre$averages.Module_num, levels = module_order_cre))) +
  geom_point( size = 3) +  # Set point color and size
  labs(title = "Scatter Plot of Avg. ENC vs dn/ds",
       x = "ENC",
       y = "dn/ds") +
  theme_minimal()+
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )

D9_IV_c = ggplot(module_stats_cre, aes(y = module_stats_cre$log_dn_ds, 
                                 x = module_stats_cre$averages.average_enc, 
                                 colour = factor(module_stats_cre$averages.Module_num, levels = module_order_cre))) +
  geom_point( size = 3) +  # Set point color and size
  labs(title = "Scatter Plot of Avg. ENC vs log dn/ds",
       x = "ENC",
       y = "log 10 dn/ds") +
  theme_minimal()+
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )

```

###V. Calculating significant differences between average dn/ds to compare modules

ANOVA and Tukey's HSD post-hoc test
```{r}
# Perform ANOVA
anova_result_cre <- aov(dn.ds_corrected ~ Module_num, data = cumulative_enc_wgcna_filt_cre)
summary(anova_result_cre)

# Tukey's HSD post-hoc test
tukey_result_cre <- TukeyHSD(anova_result_cre)
View((tukey_result_cre[["Module_num"]]))
tukey_Reslt_df_cre = as.data.frame((tukey_result_cre[["Module_num"]])) #20 significant comparisons p < 0.001

# Uncomment to write the results to text files
write.csv(tukey_Reslt_df_cre, file = "../3-Output/WGCNA/tukey_module_comparison_cre.csv",
            row.names = TRUE, col.names = FALSE, quote = FALSE)

##Plotting with p-value highlighted
#library(ggsignif)

ggplot(cumulative_enc_wgcna_filt_cre, aes(x = Module_num, y = dn.ds_corrected)) +
  geom_violin(fill = "lightblue", color = "black") +
  labs(title = "Violin plot of dN/dS by Modules",
       x = "Modules",
       y = "dN/dS") +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, 
               fill = "white", color = "black", stroke = 1) +
  # scale_x_discrete(labels = c("black" = "M4", "blue" = "M9", "brown" = "M2", 
  #                             "green" = "M6", "greenyellow" = "M3", "grey" = "M0", 
  #                             "magenta" = "M1", "pink" = "M7", "purple" = "M11", 
  #                             "red" = "M8", "turquoise" = "M5", "yellow" = "M10")) +
  geom_signif(comparisons = list(c("M8", "M4"), c("M4", "M10"))) +  # Replace with your comparisons
  theme_minimal() +
  theme(
    plot.title = element_text(size = 26, face = "bold", hjust = 0.5),
    axis.title.y = element_text(size = 18, hjust = 0.5),
    axis.title.x = element_text(size = 18),
    axis.text.x = element_text(size = 16, face = "bold", colour = "black"),
    axis.text.y = element_text(size = 16, color = "black")
    )

```

Kruskal Wallis and Dunn's test (post hoc)
```{r}

kruskal.test(cumulative_enc_wgcna_filt_cre$dn.ds_corrected ~ cumulative_enc_wgcna_filt_cre$Module_num, data = cumulative_enc_wgcna_filt_cre)


library(FSA)  # for Dunn's test
duun_test_wgcna_cre = (dunnTest(dn.ds_corrected ~ Module_num, data = cumulative_enc_wgcna_filt_cre, method = "bonferroni"))
dunn_test_wgcna_cre = as.data.frame(duun_test_wgcna_cre$res)

# Uncomment to write the results to text files
write.csv(dunn_test_wgcna_cre, file = "../3-Output/WGCNA/kruska_dunn_module_comparison_cre.csv",
            row.names = TRUE, col.names = FALSE, quote = FALSE)

```


###VI. Organizing Plots
```{r, warning=FALSE}
# Arrange plots in a single panel
combined_dn_ds_cre <- D9_I_a + D9_II_a + D9_III_a + D9_IV_b +plot_layout(ncol = 2)


# Saving as pdf
ggsave("../3-Output/WGCNA/cumulative_dnds_enc_cre.pdf", plot = combined_dn_ds_cre, width = 34, height = 15)

# #Saving as png
ggsave("../3-Output/WGCNA/cumulative_dnds_enc_cre.png", plot = combined_dn_ds_cre, width = 34, height = 15)

```

